---
title: "MIKK panel karyoplot app"
author: "Ian Brettell"
date: "25/11/2020"
output: html_document
runtime: shiny
---

```{r, include = F}
library(here)
setwd(here("shiny", "mikk_karyo"))
```


```{r, include = F}
# Setup
library(karyoploteR)
library(tidyverse)
library(BiocManager)
options(repos = BiocManager::repositories())
library(GenomicRanges)
library(biomaRt)
```


```{r, echo = F}
# Make custom Karyoplot scaffold

# Get chromosome lengths
med_chr_lens = read.table("Oryzias_latipes.ASM223467v1.dna.toplevel.fa_chr_counts.txt",
                          col.names = c("chr", "end"))
# Add start
med_chr_lens$start = 1
# Reorder
med_chr_lens = med_chr_lens %>% 
  dplyr::select(chr, start, end)
# Create custom genome
med_genome = regioneR::toGRanges(med_chr_lens)
# Create plot
#kp = karyoploteR::plotKaryotype(genome = med_genome)
```


```{r, echo = F}
# Read in ABBA-BABA sliding windows data and process

# Read in data
df = read.table("20201022_abba_sliding_windows.txt", header = T, sep = "\t", as.is = T)

# Convert fd to 0 if D < 0
df$fd = ifelse(df$D < 0,
               0,
               df$fd)

# Change names
df = df %>% 
  dplyr::mutate(p2 = recode(df$p2, hdrr = "HdrR", hni = "HNI", hsok = "HSOK"))

# Set colours
cols <- c("#F3B61F", "#631E68", "#F6673A", "#F33A56", "#55B6B0", "#08605F", "#002642", "#B02156")
names(cols) <- c("HdrR", "HSOK", "HNI", "melastigma", "javanicus", "KW", "HO5", "iCab")

# Filter for melastigma
df_kp = df %>%
  dplyr::filter(p1 == "melastigma")
# make chr numeric
df_kp$scaffold <- as.numeric(df_kp$scaffold)
```



```{r, eval = F, echo = F}
# Get `HdrR` reference exon density

# Select dataset
olat_mart = useEnsembl(biomart = "ensembl", dataset = "olatipes_gene_ensembl")
# Get attributes of interest (exon ID, chr, start, end)
exons <- getBM(attributes = c("chromosome_name", "ensembl_gene_id", "ensembl_transcript_id", "transcript_start", "transcript_end", "transcript_length", "ensembl_exon_id", "rank", "strand", "exon_chrom_start", "exon_chrom_end", "cds_start", "cds_end"),
               mart = olat_mart)
# Convert exons to GRanges
ex_ranges = GenomicRanges::makeGRangesFromDataFrame(exons,
                                                    ignore.strand = T,
                                                    seqnames.field = "chromosome_name",
                                                    start.field = "exon_chrom_start",
                                                    end.field = "exon_chrom_end")
# Save to repo
saveRDS(ex_ranges, here("data", "20201125_exon_ranges.rds"))
```

```{r, include = F, echo = F}
# Load exon density for app
ex_ranges = readRDS("20201125_exon_ranges.rds")
```


```{bash, eval = F, echo = F}
# Get MIKK SNP density

bcftools view \
  --min-alleles 2 \
  --max-alleles 2 \
  --types snps \
  --output-type u \
  vcfs/panel_no-sibs_line-ids_no-missing.vcf.gz |\
    bcftools query \
      --format '%CHROM\t%POS\n' \
      --output geno/20201023_abba_sw_granges_final/20201023_mikk.txt
      
gzip -k geno/20201023_abba_sw_granges_final/20201023_mikk.txt      
```

```{r, eval = F, echo = F}
# Read in file
mikk_ranges_df = read.table(here("geno", "20201023_abba_sw_granges_final", "20201023_mikk.txt.gz"),
                            col.names = c("chr", "pos"), 
                            sep = "\t", 
                            as.is = T)

# If on local:
#mikk_ranges_df = read.table("~/Documents/Data/20201012_mikk_genome/20201023_mikk.txt.gz",
#                            col.names = c("chr", "pos"), 
#                            sep = "\t", 
#                            as.is = T)

# Convert to GRanges object
mikk_ranges = GenomicRanges::makeGRangesFromDataFrame(mikk_ranges_df,
                                                      ignore.strand = T,
                                                      seqnames.field = "chr",
                                                      start.field = "pos",
                                                      end.field = "pos")

saveRDS(mikk_ranges, here("mikk_genome", "data", "20201125_mikk_ranges.rds"))
# If on local:
#saveRDS(mikk_ranges, here("data", "20201125_mikk_ranges.rds"))
```

```{r, include = F}
# Load MIKK SNP density for app
mikk_ranges = readRDS("20201125_mikk_ranges.rds")
```

```{bash, eval = F, echo = F}
# Get SNP density for other Oryzias 

## Get SNP density for HNI and HSOK

# Pull out HdrR, HNI and HSOK columns
mkdir geno/20201023_abba_sw_granges

for i in $( seq 1 24 ) ; do
  chr=$i ;
  in_file=geno/20201022_final_no-sibs/$chr.txt ;
  out_file=geno/20201023_abba_sw_granges/$chr.txt ;
  awk '{print $1,$2,$3,$4,$5}' $in_file \
    > $out_file ;
done

# Create final output directory
mkdir geno/20201023_abba_sw_granges/final
```

```{r, eval = F, include=F}
# Read files into R
files = list.files("geno/20201023_abba_sw_granges", full.names = T)

df = lapply(files, function(x){
  out = read.table(x,
                   header = T, 
                   as.is = T, 
                   comment.char = "*",
                   check.names = F)
  return(out)
}) %>% 
  dplyr::bind_rows() %>% 
  dplyr::select(chr = "#CHROM",
                pos = "POS",
                everything()) %>% 
  dplyr::mutate(dplyr::across(hdrr:hsok,
                              ~ifelse(.x == "N/N", NA, 1)))
  

write.table(df, 
            file = "geno/20201023_abba_sw_granges_final/20201023_hdrr_hni_hsok.txt",
            quote = F,
            sep = "\t",
            row.names = F)
```

```{bash, eval = F, include = F}
# Make version with just HNI and HSOK
awk -v OFS="\t" '{print $1, $2, $4, $5}' \
  geno/20201023_abba_sw_granges_final/20201023_hdrr_hni_hsok.txt \
  > geno/20201023_abba_sw_granges_final/20201023_hni_hsok.txt
  
gzip -k geno/20201023_abba_sw_granges_final/20201023_hni_hsok.txt  
  
# Send to local
scp brettell@ebi:/hps/research1/birney/users/ian/mikk_paper/geno/20201023_abba_sw_granges_final/20201023_hni_hsok.txt.gz ~/Documents/Data/20201012_mikk_genome
```

```{r, eval = F, include = F}
# Read in file
ol_ranges_df = read.table(here("geno", "20201023_abba_sw_granges_final", "20201023_hni_hsok.txt.gz"),
                       header = T, 
                       sep = "\t", 
                       as.is = T)

## If working on local
#ol_ranges_df = read.table("~/Documents/Data/20201012_mikk_genome/20201023_hni_hsok.txt.gz",
#                          header = T, 
#                          sep = "\t", 
#                          as.is = T)

ol_ranges_df_long = ol_ranges_df %>% 
  tidyr::pivot_longer(cols = c(hni, hsok), names_to = "line", values_to = "present")

ol_ranges_list = split(ol_ranges_df_long, f = ol_ranges_df_long$line)

ol_ranges_list = lapply(ol_ranges_list, function(x){
  # remove NAs
  df = x %>% 
    tidyr::drop_na(present)
  # convert to GRanges object
  ol_ranges = GenomicRanges::makeGRangesFromDataFrame(df,
                                                      ignore.strand = T,
                                                      seqnames.field = "chr",
                                                      start.field = "pos",
                                                      end.field = "pos")
  return(ol_ranges)
})

## Convert to GRanges object
#ol_ranges = GenomicRanges::makeGRangesFromDataFrame(ol_ranges_df,
#                                                    ignore.strand = T,
#                                                    seqnames.field = "chr",
#                                                    start.field = "pos",
#                                                    end.field = "pos")

# Save 
saveRDS(ol_ranges_list, here("data", "20201125_ol_ranges.rds"))
```

```{r, include = F}
# Load exon density for app
ol_ranges_list = readRDS("20201125_ol_ranges.rds")
```


```{r, echo = F, eval = F}
# Plot testing

inputPanel(
  selectInput("chromosome", label = "Chromosome",
              choices = seq(1, 24), selected = 2)
)

# Plot 
renderPlot({
  
  kp = plotKaryotype(med_genome, chromosomes = input$chromosome)
  # Add base numbers 
  karyoploteR::kpAddBaseNumbers(kp, tick.dist = 5000000, cex = 0.3)
  # Add data backgrounds
  karyoploteR::kpDataBackground(kp, r0=0, r1 = 1, color = "white")
  # Add axis label
  kpAxis(kp, r0=0.6, r1 = 1, cex = 0.4)
  # Add fd data
  karyoploteR::kpLines(kp,
                       chr = df_kp$scaffold[df_kp$p2 == "HNI"],
                       x = df_kp$mid[df_kp$p2 == "HNI"],
                       y = df_kp$fd[df_kp$p2 == "HNI"],
                       col = "#F6673A",
                       r0=0.6, r1 = 1)
  karyoploteR::kpLines(kp,
                       chr = df_kp$scaffold[df_kp$p2 == "HdrR"],
                       x = df_kp$mid[df_kp$p2 == "HdrR"],
                       y = df_kp$fd[df_kp$p2 == "HdrR"],
                       col = "#F3B61F",
                       r0=0.6, r1 = 1)
  karyoploteR::kpLines(kp,
                       chr = df_kp$scaffold[df_kp$p2 == "HSOK"],
                       x = df_kp$mid[df_kp$p2 == "HSOK"],
                       y = df_kp$fd[df_kp$p2 == "HSOK"],
                       col = "#631E68",
                       r0=0.6, r1 = 1)
  # Add SNP density data
  kpPlotDensity(kp, data=mikk_ranges, col = "#49A379",
                r0=0, r1=0.2, 
                window.size = 25000)
  kpPlotDensity(kp, data=ol_ranges_list$hni, col = "#F6673A",
                r0=0.2, r1=0.4, 
                window.size = 25000)
  kpPlotDensity(kp, data=ol_ranges_list$hsok, col = "#631E68", 
                r0=0.4, r1=0.6, 
                window.size = 25000)
  # Add exon density to ideogram
  kpPlotDensity(kp, data=ex_ranges, col = "#f77cb5",
                data.panel = "ideogram",
                window.size = 25000,
                r0 = 0.5, r1 = 1)
  kpPlotDensity(kp, data=ex_ranges, col = "#f77cb5",
                data.panel = "ideogram",
                window.size = 25000,
                r0 = 0.5, r1 = 0)
  # Add labels
  kpAddLabels(kp, labels="MIKK",
              r0=0, r1=0.2, 
              cex = 0.4)
  kpAddLabels(kp, labels="HNI",
              r0=0.2, r1=0.4, 
              cex = 0.4)
  kpAddLabels(kp, labels="HSOK",
              r0=0.4, r1=0.46, 
              cex = 0.4)
  kpAddLabels(kp, labels=bquote(italic(f[d])),
              r0=0.6, r1=1, 
              label.margin = 0.035,
              cex = 0.6)
  
})

```

```{r, echo = F}
library(shiny)

shiny::shinyApp(
  ui = fluidPage(
    shiny::checkboxGroupInput("chromosome", label = "Chromosome",
                choices = seq(1, 24), selected = 2),
    shiny::plotOutput("karyoplot")
  ),
  
  server = function(input, output) {
    output$karyoplot = shiny::renderPlot({
      
      kp = plotKaryotype(med_genome, chromosomes = input$chromosome)
      # Add base numbers 
      karyoploteR::kpAddBaseNumbers(kp, tick.dist = 5000000, cex = 0.3)
      # Add data backgrounds
      karyoploteR::kpDataBackground(kp, r0=0, r1 = 1, color = "white")
      # Add axis label
      kpAxis(kp, r0=0.6, r1 = 1, cex = 0.4)
      # Add fd data
      karyoploteR::kpLines(kp,
                           chr = df_kp$scaffold[df_kp$p2 == "HNI"],
                           x = df_kp$mid[df_kp$p2 == "HNI"],
                           y = df_kp$fd[df_kp$p2 == "HNI"],
                           col = "#F6673A",
                           r0=0.6, r1 = 1)
      karyoploteR::kpLines(kp,
                           chr = df_kp$scaffold[df_kp$p2 == "HdrR"],
                           x = df_kp$mid[df_kp$p2 == "HdrR"],
                           y = df_kp$fd[df_kp$p2 == "HdrR"],
                           col = "#F3B61F",
                           r0=0.6, r1 = 1)
      karyoploteR::kpLines(kp,
                           chr = df_kp$scaffold[df_kp$p2 == "HSOK"],
                           x = df_kp$mid[df_kp$p2 == "HSOK"],
                           y = df_kp$fd[df_kp$p2 == "HSOK"],
                           col = "#631E68",
                           r0=0.6, r1 = 1)
      # Add SNP density data
      kpPlotDensity(kp, data=mikk_ranges, col = "#49A379",
                    r0=0, r1=0.2, 
                    window.size = 25000)
      kpPlotDensity(kp, data=ol_ranges_list$hni, col = "#F6673A",
                    r0=0.2, r1=0.4, 
                    window.size = 25000)
      kpPlotDensity(kp, data=ol_ranges_list$hsok, col = "#631E68", 
                    r0=0.4, r1=0.6, 
                    window.size = 25000)
      # Add exon density to ideogram
      kpPlotDensity(kp, data=ex_ranges, col = "#f77cb5",
                    data.panel = "ideogram",
                    window.size = 25000,
                    r0 = 0.5, r1 = 1)
      kpPlotDensity(kp, data=ex_ranges, col = "#f77cb5",
                    data.panel = "ideogram",
                    window.size = 25000,
                    r0 = 0.5, r1 = 0)
      # Add labels
      kpAddLabels(kp, labels="MIKK",
                  r0=0, r1=0.2, 
                  cex = 0.4)
      kpAddLabels(kp, labels="HNI",
                  r0=0.2, r1=0.4, 
                  cex = 0.4)
      kpAddLabels(kp, labels="HSOK",
                  r0=0.4, r1=0.46, 
                  cex = 0.4)
      kpAddLabels(kp, labels=bquote(italic(f[d])),
                  r0=0.6, r1=1, 
                  label.margin = 0.035,
                  cex = 0.6)
      
    })  
  }
)



```


