<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>MIKK panel - LD decay analysis</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cerulean.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/anchor-sections-1.0/anchor-sections.css" rel="stylesheet" />
<script src="site_libs/anchor-sections-1.0/anchor-sections.js"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>

<link rel="stylesheet" href="styles.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">MIKK panel analysis</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="20201120_markdown_ld-decay.html">LD</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">MIKK panel - LD decay analysis</h1>

</div>


<div id="setup" class="section level1">
<h1><span class="header-section-number">1</span> Setup</h1>
<div id="create-conda-evironment" class="section level2">
<h2><span class="header-section-number">1.1</span> Create conda evironment</h2>
<pre class="bash"><code>conda env create \
  -n mikk_env \
  -f mikk_genome/code/config/conda_env.yml
  
conda activate mikk_env</code></pre>
</div>
<div id="load-r-packages" class="section level2">
<h2><span class="header-section-number">1.2</span> Load R packages</h2>
<pre class="r"><code>require(here)
require(tidyverse)</code></pre>
</div>
<div id="create-directory-structure-and-clone-repo" class="section level2">
<h2><span class="header-section-number">1.3</span> Create directory structure and clone repo</h2>
<p>(Working directory on EBI cluster: <code>/hps/research1/birney/users/ian/mikk_paper</code>)</p>
<pre class="bash"><code># move to working directory
cd /your/working/directory
# clone git repository
git clone https://github.com/Ian-Brettell/mikk_genome.git
# create directory for VCFs
mkdir vcfs</code></pre>
</div>
<div id="copy-mikk-panel-vcf-into-working-directory" class="section level2">
<h2><span class="header-section-number">1.4</span> Copy MIKK panel VCF into working directory</h2>
<p>(See supplementary material for how VCF was generated.)</p>
<pre class="bash"><code>cp /nfs/research1/birney/projects/medaka/inbred_panel/medaka-alignments-release-94/vcf/medaka_inbred_panel_ensembl_new_reference_release_94.vcf* vcfs</code></pre>
</div>
<div id="key-value-file-for-cram-id-to-line-id" class="section level2">
<h2><span class="header-section-number">1.5</span> Key-value file for cram ID to line ID</h2>
<p><code>mikk_genome/data/20200206_cram_id_to_line_id.txt</code></p>
</div>
<div id="remove-sibling-lines-and-replicates" class="section level2">
<h2><span class="header-section-number">1.6</span> Remove sibling lines and replicates</h2>
<p><strong>Full list of 80 extant MIKK panel lines</strong>: <code>mikk_genome/data/20200210_panel_lines_full.txt</code></p>
<p><strong>Note</strong>: Line <code>130-2</code> is missing from the MIKK panel VCF.</p>
<p>Identify sibling lines</p>
<pre class="bash"><code>cat mikk_genome/data/20200210_panel_lines_full.txt | cut -f1 -d&quot;-&quot; | sort | uniq -d</code></pre>
<ul>
<li>106</li>
<li>11</li>
<li>117</li>
<li>131</li>
<li>132</li>
<li>135</li>
<li>14</li>
<li>140</li>
<li>23</li>
<li>39</li>
<li>4</li>
<li>40</li>
<li>59</li>
<li>69</li>
<li>72</li>
<li>80</li>
</ul>
<p>Only keep first sibling line ( suffix _1); manually remove all others and write list of non-sibling lines to here: <code>mikk_genome/data/20200227_panel_lines_no-sibs.txt</code>. 64 lines total.</p>
<p>Excluded sibling lines here: <code>mikk_genome/data/20200227_panel_lines_excluded.txt</code>. 16 lines total.</p>
<p>Replace all dashes with underscores to match <code>mikk_genome/data/20200206_cram_id_to_line_id.txt</code> key file</p>
<pre class="bash"><code>sed &#39;s/-/_/g&#39; mikk_genome/data/20200227_panel_lines_no-sibs.txt \
  &gt; mikk_genome/data/20200227_panel_lines_no-sibs_us.txt</code></pre>
<p>Extract the lines to keep from the key file.</p>
<pre class="bash"><code>awk  &#39;FNR==NR {f1[$0]; next} $2 in f1&#39; \
  mikk_genome/data/20200227_panel_lines_no-sibs_us.txt \
  mikk_genome/data/20200206_cram_id_to_line_id.txt \
    &gt; mikk_genome/data/20200227_cram2line_no-sibs.txt</code></pre>
<p>Has 66 lines instead of 63 (64 lines minus <code>130-2</code>, which isn’t in the VCF), so there must be replicates Find out which ones:</p>
<pre class="bash"><code>cat mikk_genome/data/20200227_cram2line_no-sibs.txt | cut -f2 | cut -f1 -d&quot;_&quot; | sort | uniq -d</code></pre>
<p>32 71 84</p>
<p>Manually removed duplicate lines (<code>mikk_genome/data/20200227_duplicates_excluded.txt</code>):</p>
<ul>
<li>24271_7#5 32_2</li>
<li>24271_8#4 71_1</li>
<li>24259_1#1 84_2</li>
</ul>
<p>Final no-sibling-lines CRAM-to-lineID key file: <code>mikk_genome/data/20200227_cram2line_no-sibs.txt</code></p>
</div>
</div>
<div id="create-no-sibling-lines-mikk-panel-vcf" class="section level1">
<h1><span class="header-section-number">2</span> Create no-sibling-lines MIKK panel VCF</h1>
<pre class="bash"><code># create no-sibs file with CRAM ID only
cut -f1 mikk_genome/data/20200227_cram2line_no-sibs.txt \
  &gt; mikk_genome/data/20200227_cram2line_no-sibs_cram-only.txt
  
# make new VCF having filtered out non-MIKK and sibling lines
bcftools view \
  --output-file vcfs/panel_no-sibs.vcf \
  --samples-file mikk_genome/data/20200227_cram2line_no-sibs_cram-only.txt \
  vcfs/medaka_inbred_panel_ensembl_new_reference_release_94.vcf
  
# recode with line IDs
bcftools reheader \
  --output vcfs/panel_no-sibs_line-ids.vcf \
  --samples mikk_genome/data/20200227_cram2line_no-sibs.txt \
  vcfs/panel_no-sibs.vcf
  
# compress
bcftools view \
  --output-type z \
  --output-file vcfs/panel_no-sibs_line-ids.vcf.gz \
  vcfs/panel_no-sibs_line-ids.vcf
  
# index
bcftools index \
  --tbi \
  vcfs/panel_no-sibs_line-ids.vcf.gz

# get stats
mkdir stats

bcftools stats \
  vcfs/panel_no-sibs_line-ids.vcf.gz \
  &gt; stats/20200305_panel_no-sibs.txt

## get basic counts
grep &quot;^SN&quot; stats/20200305_panel_no-sibs.txt</code></pre>
<div id="make-a-version-with-no-missing-variants" class="section level2">
<h2><span class="header-section-number">2.1</span> Make a version with no missing variants</h2>
<pre class="bash"><code>vcftools \
  --gzvcf vcfs/panel_no-sibs_line-ids.vcf.gz \
  --max-missing 1 \
  --recode \
  --stdout &gt; vcfs/panel_no-sibs_line-ids_no-missing.vcf
  
# compress
bcftools view \
  --output-type z \
  --output-file vcfs/panel_no-sibs_line-ids_no-missing.vcf.gz \
  vcfs/panel_no-sibs_line-ids_no-missing.vcf

# create index
bcftools index \
  --tbi vcfs/panel_no-sibs_line-ids_no-missing.vcf.gz
  
# get stats 
bcftools stats \
  vcfs/panel_no-sibs_line-ids_no-missing.vcf.gz \
  &gt; stats/20200305_panel_no-sibs_no-missing.txt

# get basic counts
grep &quot;^SN&quot; stats/20200305_panel_no-sibs_no-missing.txt</code></pre>
</div>
</div>
<div id="generate-haploview-plots" class="section level1">
<h1><span class="header-section-number">3</span> Generate Haploview plots</h1>
<div id="create-plink-dataset-from-no-sib-lines-no-missing-vcf" class="section level2">
<h2><span class="header-section-number">3.1</span> Create <code>plink</code> dataset from no-sib-lines, no-missing VCF</h2>
<pre class="bash"><code>mkdir plink/20200716_panel_no-sibs_line-ids_no-missing

# make BED  
plink \
  --vcf vcfs/panel_no-sibs_line-ids_no-missing.vcf.gz \
  --make-bed \
  --double-id \
  --snps-only \
  --biallelic-only \
  --chr-set 24 no-xy \
  --chr 1-24 \
  --out plink/20200716_panel_no-sibs_line-ids_no-missing/20200716
  
# recode for 012 transposed
plink \
  --bfile plink/20200716_panel_no-sibs_line-ids_no-missing/20200716 \
  --recode A-transpose \
  --out plink/20200716_panel_no-sibs_line-ids_no-missing/20200716_recode012</code></pre>
</div>
<div id="create-bed-sets-filtered-for-maf-0.03-0.05-and-0.10" class="section level2">
<h2><span class="header-section-number">3.2</span> Create BED sets filtered for MAF &gt; 0.03, 0.05 and 0.10</h2>
<pre class="bash"><code>maf_thresholds=$( echo 0.03 0.05 0.10 )

# Make new BEDs 
for i in $maf_thresholds ; do
  # make directory
  new_path=plink/20200716_panel_no-sibs_line-ids_no-missing/20200803_maf-$i ;
  # make directory
  if [ ! -d &quot;$new_path&quot; ]; then
    mkdir $new_path;
  fi
  # make BED set
  plink \
    --bfile plink/20200716_panel_no-sibs_line-ids_no-missing/20200716 \
    --make-bed \
    --double-id \
    --chr-set 24 no-xy \
    --maf $i \
    --out $new_path/20200803
done</code></pre>
</div>
<div id="recode-for-haploview" class="section level2">
<h2><span class="header-section-number">3.3</span> Recode for Haploview</h2>
<pre class="bash"><code># Create output directory
mkdir plink/20200716_panel_no-sibs_line-ids_no-missing/20200803_hv_thinned

hv_thinned_path=plink/20200716_panel_no-sibs_line-ids_no-missing/20200803_hv_thinned

# Recode
for i in $maf_thresholds ; do
  new_path=$hv_thinned_path/$i ;
  # make directory
  if [ ! -d &quot;$new_path&quot; ]; then
    mkdir $new_path;
  fi 
  # recode 
  for j in $(seq 1 24); do
    plink \
      --bfile plink/20200716_panel_no-sibs_line-ids_no-missing/20200803_maf-$i/20200803 \
      --recode HV-1chr \
      --double-id \
      --chr-set 24 no-xy \
      --chr $j \
      --allele1234 \
      --thin-count 3000 \
      --out $hv_thinned_path/$i/20200803_chr-$j;
  done;
done

# Edit .ped files to remove asterisks
for i in $maf_thresholds ; do
  for j in $(find $hv_thinned_path/$i/20200803_chr-*.ped); do
    sed -i &#39;s/\*/0/g&#39; $j;
  done;
done  

# Edit .info files to make the SNP&#39;s bp position its ID
for i in $maf_thresholds; do
  for j in $(find $hv_thinned_path/$i/20200803_chr*.info); do
    outname=$(echo $j\_with-id);
    awk -v OFS=&quot;\t&quot; {&#39;print $2,$2&#39;} $j &gt; $outname;
  done;
done</code></pre>
</div>
<div id="plot" class="section level2">
<h2><span class="header-section-number">3.4</span> Plot</h2>
<p><strong>NOTE</strong>: This code requires <code>Haploview</code>, which you will need to install on your system: <a href="https://www.broadinstitute.org/haploview/haploview" class="uri">https://www.broadinstitute.org/haploview/haploview</a></p>
<pre class="bash"><code>hv_path=/nfs/software/birney/Haploview.jar # edit to your Haploview path

mkdir plots/20200803_ld_thinned/

for i in $maf_thresholds; do
  # set output directory
  new_path=plots/20200803_ld_thinned/$i ;
  # make directory
  if [ ! -d &quot;$new_path&quot; ]; then
    mkdir $new_path;
  fi   
  for j in $(seq 1 24); do
    bsub -M 20000 -o log/20200803_hv_$i\_$j.out -e log/20200803_hv_$i\_$j.err \
    &quot;java -Xms18G -Xmx18G -jar $hv_path \
      -memory 18000 \
      -pedfile $hv_thinned_path/$i/20200803_chr-$j.ped  \
      -info $hv_thinned_path/$i/20200803_chr-$j.info_with-id \
      -maxDistance 1000 \
      -ldcolorscheme DEFAULT \
      -ldvalues RSQ \
      -minMAF $i \
      -nogui \
      -svg \
      -out $new_path/$j&quot;;
  done;
done</code></pre>
<p>These <code>svg</code> files can be converted to <code>pdf</code> using: * <a href="https://www.zamzar.com/" class="uri">https://www.zamzar.com/</a> for files &gt; 30 MB (chr 1) - note limit on number of files you can convert * <a href="https://onlineconvertfree.com/convert-format/svg-to-pdf/" class="uri">https://onlineconvertfree.com/convert-format/svg-to-pdf/</a> for the rest</p>
<p>The full Haploview LD plots are available in the Supplementary Material.</p>
<p>By inspecting these LD plots at the <code>MAF &gt; 0.05</code> level, we discovered the following LD blocks worthy of further investigation:</p>
<ul>
<li>5:28181970-28970558 (788 Kb)</li>
<li>6:29398579-32246747 (2.85 Mb)</li>
<li>12:25336174-25384053 (48 Kb)</li>
<li>14:12490842-12947083 (456 Kb)</li>
<li>17:15557892-19561518 (4 Mb)</li>
<li>21:6710074-7880374 (1.17 Mb)</li>
</ul>
<p>See zoomed plots here:</p>
<div class="figure">
<img src="hv_5_28181970-28970558.png" alt="" />
<p class="caption"><strong>5:28181970-28970558</strong></p>
</div>
<div class="figure">
<img src="hv_6_29398579-32246747.png" alt="" />
<p class="caption"><strong>6:29398579-32246747</strong></p>
</div>
<p><img src="hv_12_25336174-25384053.png" alt="12:25336174-25384053" /> <img src="hv_14_12490842-12947083.png" alt="14:12490842-129470833" /></p>
<div class="figure">
<img src="hv_17_15557892-19561518.png" alt="" />
<p class="caption"><strong>17:15557892-19561518</strong></p>
</div>
<div class="figure">
<img src="hv_21_6710074-7880374.png" alt="" />
<p class="caption"><strong>hv_21_6710074-7880374</strong></p>
</div>
</div>
</div>
<div id="ld-decay" class="section level1">
<h1><span class="header-section-number">4</span> LD decay</h1>
<div id="obtain" class="section level2">
<h2><span class="header-section-number">4.1</span> Obtain</h2>
<pre class="bash"><code># make BED
mkdir plink/20200727_mikk_no-missing_maf-0.05

plink \
  --vcf vcfs/panel_no-sibs_line-ids.vcf.gz \
  --make-bed \
  --double-id \
  --snps-only \
  --biallelic-only \
  --maf 0.05 \
  --geno 0 \
  --chr-set 24 no-xy \
  --out plink/20200727_mikk_no-missing_maf-0.05/20200727

# get LD   
mkdir ld/20200727_mikk_maf-0.10_window-50kb_no-missing/

for i in $(seq 1 24); do
  plink \
      --bfile plink/20200727_mikk_no-missing_maf-0.05/20200727 \
      --r2 \
      --ld-window 999999 \
      --ld-window-kb 50 \
      --ld-window-r2 0 \
      --chr-set 24 no-xy \
      --chr $i \
      --maf 0.10 \
      --out ld/20200727_mikk_maf-0.10_window-50kb_no-missing/$i;
done

# for 1KG too
mkdir ld/20200727_1kg_maf-0.10_window-50kb_no-missing/

for i in $(seq 1 22); do
  plink \
      --bfile plink/20200723_1gk_no-missing_maf-0.05/20200723 \
      --r2 \
      --ld-window 999999 \
      --ld-window-kb 50 \
      --ld-window-r2 0 \
      --chr $i \
      --maf 0.10 \
      --out ld/20200727_1kg_maf-0.10_window-50kb_no-missing/$i;
done

# do again with ld-window-kb 10 to get counts of comparisons for paper
# get LD   
mkdir ld/20200803_mikk_maf-0.10_window-10kb_no-missing/

for i in $(seq 1 24); do
  plink \
      --bfile plink/20200727_mikk_no-missing_maf-0.05/20200727 \
      --r2 \
      --ld-window 999999 \
      --ld-window-kb 10 \
      --ld-window-r2 0 \
      --chr-set 24 no-xy \
      --chr $i \
      --maf 0.10 \
      --out ld/20200803_mikk_maf-0.10_window-10kb_no-missing/$i;
done

# for 1KG too
mkdir ld/20200803_1kg_maf-0.10_window-10kb_no-missing/

for i in $(seq 1 22); do
  plink \
      --bfile plink/20200723_1gk_no-missing_maf-0.05/20200723 \
      --r2 \
      --ld-window 999999 \
      --ld-window-kb 10 \
      --ld-window-r2 0 \
      --chr $i \
      --maf 0.10 \
      --out ld/20200803_1kg_maf-0.10_window-10kb_no-missing/$i;
done

# Get counts
wc -l ld/20200803_mikk_maf-0.10_window-10kb_no-missing/*.ld
# Total: 
wc -l ld/20200803_1kg_maf-0.10_window-10kb_no-missing/*.ld</code></pre>
</div>
</div>

<p>Copyright &copy; 2016 Skynet, Inc. All rights reserved.</p>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
