<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />



<meta name="date" content="2021-03-18" />

<title>Introgression</title>

<script src="site_libs/header-attrs-2.7/header-attrs.js"></script>
<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/flatly.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>


<style type="text/css">
  code {
    white-space: pre;
  }
  .sourceCode {
    overflow: visible;
  }
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>







<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-sm-12 col-md-4 col-lg-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-sm-12 col-md-8 col-lg-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">MIKK panel genome</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="20210301_fecundity.html">Fecundity</a>
</li>
<li>
  <a href="20210305_ld-decay.html">Linkage Disequilibrium</a>
</li>
<li>
  <a href="20210318_introgression.html">Introgression</a>
</li>
<li>
  <a href="20210301_repeats.html">Repeat Elements</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Introgression</h1>
<h4 class="date">2021-03-18</h4>

</div>


<p>We want to determine the extent to which the MIKK panel shows evidence of introgression with other medaka populations, specifically the Northern and Korean medaka strains.</p>
<div id="setup" class="section level1" number="1">
<h1 number="1"><span class="header-section-number">1</span> Setup</h1>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(here)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">source</span>(here<span class="sc">::</span><span class="fu">here</span>(<span class="st">&quot;code&quot;</span>, <span class="st">&quot;scripts&quot;</span>, <span class="st">&quot;introgression&quot;</span>, <span class="st">&quot;source.R&quot;</span>))</span></code></pre></div>
<div id="working-directory" class="section level2" number="1.1">
<h2 number="1.1"><span class="header-section-number">1.1</span> Working directory</h2>
<p>Working directory on EBI cluster: <code>/hps/research1/birney/users/ian/mikk_paper</code></p>
</div>
<div id="create-singularity-containers" class="section level2" number="1.2">
<h2 number="1.2"><span class="header-section-number">1.2</span> Create Singularity containers</h2>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="va">container_dir=</span>../sing_conts</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="co"># load Singularity (version 3.5.0)</span></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="ex">module</span> load singularity </span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> package <span class="kw">in</span> <span class="va">$(</span> <span class="bu">echo</span> r-base_4.0.4 bcftools_1.9 numpy_1.15.4 bash_3.0.22 bioconductor_3.12 <span class="va">)</span> <span class="kw">;</span> <span class="cf">do</span></span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> <span class="kw">[[</span> <span class="ot">!</span> <span class="ot">-f</span> <span class="va">$container_dir</span>/<span class="va">$package</span>.sif <span class="kw">]];</span> <span class="cf">then</span></span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a>    <span class="ex">singularity</span> build <span class="dt">\</span></span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a>      <span class="at">--remote</span> <span class="dt">\</span></span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a>      <span class="va">$container_dir</span>/<span class="va">$package</span>.sif <span class="dt">\</span></span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a>      envs/<span class="va">$package</span>.def </span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a>  <span class="cf">fi</span> <span class="kw">;</span></span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a><span class="cf">done</span></span></code></pre></div>
</div>
<div id="renv" class="section level2" number="1.3">
<h2 number="1.3"><span class="header-section-number">1.3</span> <code>renv</code></h2>
<p>Install all required packages (using r-base in Singularity container).</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="va">container_dir=</span>../sing_conts</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="va">baseR=</span>r-base_4.0.4</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a><span class="ex">bsub</span> <span class="at">-Is</span> <span class="st">&quot;singularity shell </span><span class="va">$container_dir</span><span class="st">/</span><span class="va">$baseR</span><span class="st">&quot;</span></span></code></pre></div>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Install all required packages</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>renv<span class="sc">::</span><span class="fu">restore</span>()</span></code></pre></div>
</div>
<div id="copy-scripts-from-simon-martins-github-repo" class="section level2" number="1.4">
<h2 number="1.4"><span class="header-section-number">1.4</span> Copy scripts from Simon Martin’s GitHub repo</h2>
<div class="sourceCode" id="cb5"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> <span class="at">-P</span> code/scripts/introgression https://raw.githubusercontent.com/simonhmartin/genomics_general/master/ABBABABAwindows.py</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> <span class="at">-P</span> code/scripts/introgression https://raw.githubusercontent.com/simonhmartin/genomics_general/master/genomics.py</span></code></pre></div>
</div>
</div>
<div id="process-multiple-alignment-data" class="section level1" number="2">
<h1 number="2"><span class="header-section-number">2</span> Process multiple alignment data</h1>
<div id="download-from-ensembl" class="section level2" number="2.1">
<h2 number="2.1"><span class="header-section-number">2.1</span> Download from Ensembl</h2>
<ul>
<li><span style="color: purple;">Ensembl 103</span>: Only includes Oryzias latipes HdrR and Oryzias melastigma (Indian medaka).</li>
<li><span style="color: &quot;purple;">Ensembl 102</span>: Includes all target species. Use.</li>
</ul>
<p><code>ftp://ftp.ensembl.org/pub/release-102/emf/ensembl-compara/multiple_alignments/50_fish.epo/README.50_fish.epo</code> reads:</p>
<blockquote>
<p>Alignments are grouped by japanese medaka hdrr chromosome, and then by coordinate system. Alignments containing duplications in japanese medaka hdrr are dumped once per duplicated segment. The files named <em>.other</em>.emf contain alignments that do not include any japanese medaka hdrr region. Each file contains up to 200 alignments.</p>
</blockquote>
<div class="sourceCode" id="cb6"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="va">ftp_dir=</span>ftp://ftp.ensembl.org/pub/release-102/emf/ensembl-compara/multiple_alignments/50_fish.epo/</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="va">target_dir=</span>../introgression/release-102</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> <span class="va">$target_dir</span>/raw</span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a><span class="co"># download, exlcuding *.other* files</span></span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> <span class="at">-P</span> <span class="va">$target_dir</span>/raw <span class="va">$ftp_dir</span>/<span class="pp">*</span> <span class="at">-R</span> <span class="st">&quot;*other*&quot;</span></span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a><span class="co"># unzip into new directory (excluding &quot;other&quot;)</span></span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true" tabindex="-1"></a><span class="va">$target_dir</span><span class="ex">/unzipped</span></span>
<span id="cb6-11"><a href="#cb6-11" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> <span class="va">$target_dir</span>/unzipped</span>
<span id="cb6-12"><a href="#cb6-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-13"><a href="#cb6-13" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="va">$(</span><span class="fu">find</span> <span class="va">$target_dir</span>/raw/50_fish.epo.[0-9]<span class="pp">*</span><span class="va">)</span><span class="kw">;</span> <span class="cf">do</span></span>
<span id="cb6-14"><a href="#cb6-14" aria-hidden="true" tabindex="-1"></a>  <span class="va">name=$(</span><span class="fu">basename</span> <span class="va">$i</span> <span class="kw">|</span> <span class="fu">cut</span> <span class="at">-f3,4</span> <span class="at">-d</span><span class="st">&#39;.&#39;</span><span class="va">)</span><span class="kw">;</span></span>
<span id="cb6-15"><a href="#cb6-15" aria-hidden="true" tabindex="-1"></a>  <span class="ex">bsub</span> <span class="st">&quot;zcat </span><span class="va">$i</span><span class="st"> &gt; </span><span class="va">$target_dir</span><span class="st">/unzipped/</span><span class="va">$name</span><span class="st">&quot;</span><span class="kw">;</span></span>
<span id="cb6-16"><a href="#cb6-16" aria-hidden="true" tabindex="-1"></a><span class="cf">done</span></span></code></pre></div>
<ul>
<li><span style="color: red;"><strong>NOTE</strong></span>: File 6_2.emf is in a completely different format, with CIGAR strings instead of the normal SEQ, TREE, ID and DATA segments. It appears the file is corrupted.</li>
<li>The 6_2 file in <code>release 101</code> is unaffected. Remove all <code>release 102</code> chr 6 files and replace with <code>release 101</code> files.</li>
</ul>
<div class="sourceCode" id="cb7"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="co"># remove release 102 files for chr 6</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span> <span class="va">$target_dir</span>/unzipped/6_<span class="pp">*</span></span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a><span class="co"># download chr 6 files from release 101</span></span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a><span class="va">ftp_dir_101=</span>ftp://ftp.ensembl.org/pub/release-101/emf/ensembl-compara/multiple_alignments/50_fish.epo/</span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a><span class="va">target_dir_101=</span>../introgression/release-101</span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> <span class="va">$target_dir_101</span>/raw</span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-10"><a href="#cb7-10" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> <span class="at">-P</span> <span class="va">$target_dir_101</span>/raw <span class="va">$ftp_dir_101</span>/50_fish.epo.6_<span class="pp">*</span></span>
<span id="cb7-11"><a href="#cb7-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-12"><a href="#cb7-12" aria-hidden="true" tabindex="-1"></a><span class="co"># unzip</span></span>
<span id="cb7-13"><a href="#cb7-13" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> <span class="va">$target_dir_101</span>/unzipped</span>
<span id="cb7-14"><a href="#cb7-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-15"><a href="#cb7-15" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="va">$(</span><span class="fu">find</span> <span class="va">$target_dir_101</span>/raw/<span class="pp">*</span><span class="va">)</span><span class="kw">;</span> <span class="cf">do</span></span>
<span id="cb7-16"><a href="#cb7-16" aria-hidden="true" tabindex="-1"></a>  <span class="va">name=$(</span><span class="fu">basename</span> <span class="va">$i</span> <span class="kw">|</span> <span class="fu">cut</span> <span class="at">-f3,4</span> <span class="at">-d</span><span class="st">&#39;.&#39;</span><span class="va">)</span> <span class="kw">;</span></span>
<span id="cb7-17"><a href="#cb7-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">zcat</span> <span class="va">$i</span> <span class="op">&gt;</span> <span class="va">$target_dir_101</span>/unzipped/<span class="va">$name</span> <span class="kw">;</span></span>
<span id="cb7-18"><a href="#cb7-18" aria-hidden="true" tabindex="-1"></a><span class="cf">done</span></span>
<span id="cb7-19"><a href="#cb7-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-20"><a href="#cb7-20" aria-hidden="true" tabindex="-1"></a><span class="co"># copy over to release 102 directory</span></span>
<span id="cb7-21"><a href="#cb7-21" aria-hidden="true" tabindex="-1"></a><span class="fu">cp</span> <span class="va">$target_dir_101</span>/unzipped/<span class="pp">*</span> <span class="va">$target_dir</span>/unzipped</span></code></pre></div>
</div>
<div id="generate-tree-plot" class="section level2" number="2.2">
<h2 number="2.2"><span class="header-section-number">2.2</span> Generate tree plot</h2>
<p>Copy tree to file</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="va">tree_file=</span>data/introgression/release_102_tree.txt</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a><span class="fu">awk</span> <span class="st">&quot;NR==58,NR==205&quot;</span> <span class="va">$target_dir</span>/raw/README.50_fish.epo <span class="dt">\</span></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>  <span class="op">&gt;</span> <span class="va">$tree_file</span></span></code></pre></div>
<p>Then manually edit <code>$tree_file</code> using regex to find spaces and replace them with "_": <code>{bash} (?&lt;=[a-z])( )(?=[a-z])</code></p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>phylo_tree <span class="ot">&lt;-</span> ape<span class="sc">::</span><span class="fu">read.tree</span>(<span class="at">file =</span> here<span class="sc">::</span><span class="fu">here</span>(<span class="st">&quot;data&quot;</span>, <span class="st">&quot;introgression&quot;</span>, <span class="st">&quot;release_102_tree.txt&quot;</span>))</span></code></pre></div>
<div id="full-tree" class="section level3" number="2.2.1">
<h3 number="2.2.1"><span class="header-section-number">2.2.1</span> Full tree</h3>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Colour all Oryzias</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>ids <span class="ot">&lt;-</span> phylo_tree<span class="sc">$</span>tip.label[<span class="fu">grep</span>(<span class="st">&quot;Oryzias&quot;</span>, phylo_tree<span class="sc">$</span>tip.label)]</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a><span class="co"># get indices of edges descending from MRCA (determined through trial and error)</span></span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>oryzias_nodes <span class="ot">&lt;-</span> <span class="fu">seq</span>(<span class="dv">35</span>, <span class="dv">42</span>)</span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>all_med_col <span class="ot">&lt;-</span> <span class="fu">ifelse</span>(<span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(phylo_tree[[<span class="st">&quot;edge.length&quot;</span>]]) <span class="sc">%in%</span> oryzias_nodes, <span class="st">&quot;#E84141&quot;</span>, <span class="st">&quot;black&quot;</span>)</span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a><span class="co"># set colours for tip labels</span></span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a>all_med_tip <span class="ot">&lt;-</span> <span class="fu">ifelse</span>(phylo_tree<span class="sc">$</span>tip.label <span class="sc">%in%</span> ids, <span class="st">&quot;#E84141&quot;</span>, <span class="st">&quot;black&quot;</span>)</span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a><span class="co"># plot</span></span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a>ape<span class="sc">::</span><span class="fu">plot.phylo</span>(phylo_tree,</span>
<span id="cb10-10"><a href="#cb10-10" aria-hidden="true" tabindex="-1"></a>                <span class="at">use.edge.length =</span> T,</span>
<span id="cb10-11"><a href="#cb10-11" aria-hidden="true" tabindex="-1"></a>                <span class="at">edge.color =</span> all_med_col,</span>
<span id="cb10-12"><a href="#cb10-12" aria-hidden="true" tabindex="-1"></a>                <span class="at">tip.color =</span> all_med_tip,</span>
<span id="cb10-13"><a href="#cb10-13" aria-hidden="true" tabindex="-1"></a>                <span class="at">font =</span> <span class="dv">4</span>)</span></code></pre></div>
<p><img src="20210318_introgression_files/figure-html/unnamed-chunk-10-1.svg" width="1056" /></p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Save to repo</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a><span class="fu">png</span>(<span class="at">file=</span> <span class="fu">file.path</span>(plots_dir, <span class="st">&quot;tree_all.png&quot;</span>),</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">width=</span><span class="dv">22</span>,</span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">height=</span><span class="dv">25</span>,</span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">units =</span> <span class="st">&quot;cm&quot;</span>,</span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">res =</span> <span class="dv">400</span>)</span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a>ape<span class="sc">::</span><span class="fu">plot.phylo</span>(phylo_tree,</span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true" tabindex="-1"></a>                <span class="at">use.edge.length =</span> T,</span>
<span id="cb11-9"><a href="#cb11-9" aria-hidden="true" tabindex="-1"></a>                <span class="at">edge.color =</span> all_med_col,</span>
<span id="cb11-10"><a href="#cb11-10" aria-hidden="true" tabindex="-1"></a>                <span class="at">tip.color =</span> all_med_tip,</span>
<span id="cb11-11"><a href="#cb11-11" aria-hidden="true" tabindex="-1"></a>                <span class="at">font =</span> <span class="dv">4</span>)</span>
<span id="cb11-12"><a href="#cb11-12" aria-hidden="true" tabindex="-1"></a><span class="fu">dev.off</span>()</span></code></pre></div>
</div>
<div id="oryzias-only" class="section level3" number="2.2.2">
<h3 number="2.2.2"><span class="header-section-number">2.2.2</span> Oryzias only</h3>
<p>New tree file created manually to extract <em>Oryzias</em> fishes only, and replace reference codes (e.g. “ASM223467v1”) with line names (e.g. “HdrR”).</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>in_file <span class="ot">=</span> here<span class="sc">::</span><span class="fu">here</span>(<span class="st">&quot;data/introgression/release_102_tree_oryzias_only.txt&quot;</span>)</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Read in</span></span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>phylo_tree <span class="ot">&lt;-</span> ape<span class="sc">::</span><span class="fu">read.tree</span>(<span class="at">file =</span> in_file)</span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Set colours </span></span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a>phylo_cols <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;#55b6b0&quot;</span>, <span class="st">&quot;#f33a56&quot;</span>, <span class="st">&quot;#f3b61f&quot;</span>, <span class="st">&quot;#f6673a&quot;</span>, <span class="st">&quot;#631e68&quot;</span>)</span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot</span></span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a>ape<span class="sc">::</span><span class="fu">plot.phylo</span>(phylo_tree,</span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a>                <span class="at">font =</span> <span class="dv">4</span>,</span>
<span id="cb12-9"><a href="#cb12-9" aria-hidden="true" tabindex="-1"></a>                <span class="at">tip.color =</span> phylo_cols)</span></code></pre></div>
<p><img src="20210318_introgression_files/figure-html/unnamed-chunk-12-1.svg" width="672" /></p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>out_file <span class="ot">=</span> here<span class="sc">::</span><span class="fu">here</span>(plots_dir, <span class="st">&quot;tree_oryzias.png&quot;</span>)</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Save</span></span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a><span class="fu">png</span>(<span class="at">file=</span>out_file,</span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">width=</span><span class="dv">2700</span>,</span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">height=</span><span class="dv">1720</span>,</span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">units =</span> <span class="st">&quot;px&quot;</span>,</span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">res =</span> <span class="dv">400</span>)</span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a>ape<span class="sc">::</span><span class="fu">plot.phylo</span>(phylo_tree,</span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true" tabindex="-1"></a>                <span class="at">font =</span> <span class="dv">4</span>,</span>
<span id="cb13-10"><a href="#cb13-10" aria-hidden="true" tabindex="-1"></a>                <span class="at">tip.color =</span> phylo_cols)</span>
<span id="cb13-11"><a href="#cb13-11" aria-hidden="true" tabindex="-1"></a><span class="fu">dev.off</span>()</span></code></pre></div>
<div id="add-cross-for-ancestor" class="section level4" number="2.2.2.1">
<h4 number="2.2.2.1"><span class="header-section-number">2.2.2.1</span> Add cross for ancestor</h4>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>in_file <span class="ot">=</span> here<span class="sc">::</span><span class="fu">here</span>(plots_dir, <span class="st">&quot;tree_oryzias.png&quot;</span>)</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>out_file <span class="ot">=</span> here<span class="sc">::</span><span class="fu">here</span>(plots_dir, <span class="st">&quot;tree_oryzias_with_ancestor.png&quot;</span>)</span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a>tree_path <span class="ot">=</span> in_file</span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a><span class="fu">ggdraw</span>() <span class="sc">+</span></span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">draw_image</span>(tree_path) <span class="sc">+</span></span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">draw_label</span>(<span class="st">&quot;x&quot;</span>,</span>
<span id="cb14-9"><a href="#cb14-9" aria-hidden="true" tabindex="-1"></a>             <span class="at">x =</span> <span class="fl">0.152</span>,</span>
<span id="cb14-10"><a href="#cb14-10" aria-hidden="true" tabindex="-1"></a>             <span class="at">y =</span> <span class="fl">0.52</span>,</span>
<span id="cb14-11"><a href="#cb14-11" aria-hidden="true" tabindex="-1"></a>             <span class="at">fontface =</span> <span class="st">&quot;bold&quot;</span>,</span>
<span id="cb14-12"><a href="#cb14-12" aria-hidden="true" tabindex="-1"></a>             <span class="at">color =</span> <span class="st">&quot;#f77cb5&quot;</span>,</span>
<span id="cb14-13"><a href="#cb14-13" aria-hidden="true" tabindex="-1"></a>             <span class="at">size =</span> <span class="dv">25</span>)</span>
<span id="cb14-14"><a href="#cb14-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-15"><a href="#cb14-15" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(out_file,</span>
<span id="cb14-16"><a href="#cb14-16" aria-hidden="true" tabindex="-1"></a>    <span class="at">width=</span><span class="fl">15.69</span>,</span>
<span id="cb14-17"><a href="#cb14-17" aria-hidden="true" tabindex="-1"></a>    <span class="at">height=</span><span class="dv">10</span>,</span>
<span id="cb14-18"><a href="#cb14-18" aria-hidden="true" tabindex="-1"></a>    <span class="at">units =</span> <span class="st">&quot;cm&quot;</span>,</span>
<span id="cb14-19"><a href="#cb14-19" aria-hidden="true" tabindex="-1"></a>    <span class="at">dpi =</span> <span class="dv">400</span>)</span></code></pre></div>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">include_graphics</span>(<span class="fu">basename</span>(out_file))</span></code></pre></div>
<p><embed src="20210208_semiquant.csv" /><!-- --></p>
</div>
</div>
</div>
<div id="divide-by-segment" class="section level2" number="2.3">
<h2 number="2.3"><span class="header-section-number">2.3</span> Divide by segment</h2>
<div class="sourceCode" id="cb16"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="va">target_dir=</span>../introgression/release-102</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a><span class="va">segments_dir=$target_dir</span>/segmented</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a><span class="va">date=</span>20210312</span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a><span class="va">script=</span>code/scripts/introgression/20200907_extract-emf-segments.sh</span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> <span class="va">$segments_dir</span></span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-8"><a href="#cb16-8" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="va">$(</span><span class="fu">find</span> <span class="va">$target_dir</span>/unzipped/<span class="pp">*</span> <span class="va">)</span><span class="kw">;</span> <span class="cf">do</span></span>
<span id="cb16-9"><a href="#cb16-9" aria-hidden="true" tabindex="-1"></a>  <span class="co"># get basename</span></span>
<span id="cb16-10"><a href="#cb16-10" aria-hidden="true" tabindex="-1"></a>  <span class="va">bname=$(</span><span class="fu">basename</span> <span class="va">$i)</span><span class="kw">;</span></span>
<span id="cb16-11"><a href="#cb16-11" aria-hidden="true" tabindex="-1"></a>  <span class="va">bname_short=$(</span><span class="bu">echo</span> <span class="va">${bname</span><span class="op">::</span>-<span class="dv">4</span><span class="va">}</span> <span class="va">)</span><span class="kw">;</span></span>
<span id="cb16-12"><a href="#cb16-12" aria-hidden="true" tabindex="-1"></a>  <span class="co"># get chromosome</span></span>
<span id="cb16-13"><a href="#cb16-13" aria-hidden="true" tabindex="-1"></a>  <span class="va">chr=$(</span><span class="bu">echo</span> <span class="va">$bname</span> <span class="kw">|</span> <span class="fu">cut</span> <span class="at">-f1</span> <span class="at">-d</span><span class="st">&quot;_&quot;</span> <span class="va">)</span><span class="kw">;</span></span>
<span id="cb16-14"><a href="#cb16-14" aria-hidden="true" tabindex="-1"></a>  <span class="co"># make directory for each EMF file</span></span>
<span id="cb16-15"><a href="#cb16-15" aria-hidden="true" tabindex="-1"></a>  <span class="va">new_path=$(</span><span class="bu">echo</span> <span class="va">$segments_dir</span>/<span class="va">$bname_short</span> <span class="va">)</span><span class="kw">;</span></span>
<span id="cb16-16"><a href="#cb16-16" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> <span class="bu">[</span> <span class="ot">!</span> <span class="ot">-d</span> <span class="st">&quot;</span><span class="va">$new_path</span><span class="st">&quot;</span> <span class="bu">]</span><span class="kw">;</span> <span class="cf">then</span></span>
<span id="cb16-17"><a href="#cb16-17" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mkdir</span> <span class="va">$new_path</span><span class="kw">;</span></span>
<span id="cb16-18"><a href="#cb16-18" aria-hidden="true" tabindex="-1"></a>  <span class="cf">fi</span></span>
<span id="cb16-19"><a href="#cb16-19" aria-hidden="true" tabindex="-1"></a>  <span class="co"># get segment count</span></span>
<span id="cb16-20"><a href="#cb16-20" aria-hidden="true" tabindex="-1"></a>  <span class="va">segment_count=$(</span><span class="fu">grep</span> <span class="st">&quot;^DATA&quot;</span> <span class="va">$i</span> <span class="kw">|</span> <span class="fu">wc</span> <span class="at">-l</span> <span class="va">)</span><span class="kw">;</span></span>
<span id="cb16-21"><a href="#cb16-21" aria-hidden="true" tabindex="-1"></a>  <span class="co"># get segment start and end for each file</span></span>
<span id="cb16-22"><a href="#cb16-22" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> j <span class="kw">in</span> <span class="va">$(</span><span class="fu">seq</span> 1 <span class="va">$segment_count</span> <span class="va">)</span><span class="kw">;</span> <span class="cf">do</span></span>
<span id="cb16-23"><a href="#cb16-23" aria-hidden="true" tabindex="-1"></a>    <span class="ex">bsub</span> <span class="dt">\</span></span>
<span id="cb16-24"><a href="#cb16-24" aria-hidden="true" tabindex="-1"></a>      <span class="at">-o</span> ../log/<span class="va">$date</span><span class="dt">\s</span>egment_<span class="va">$bname_short</span><span class="dt">\_</span><span class="va">$j</span>.out <span class="dt">\</span></span>
<span id="cb16-25"><a href="#cb16-25" aria-hidden="true" tabindex="-1"></a>      <span class="at">-e</span> ../log/<span class="va">$date</span><span class="dt">\s</span>egment_<span class="va">$bname_short</span><span class="dt">\_</span><span class="va">$j</span>.err <span class="dt">\</span></span>
<span id="cb16-26"><a href="#cb16-26" aria-hidden="true" tabindex="-1"></a>      <span class="st">&quot;</span><span class="va">$script</span><span class="st"> </span><span class="va">$i</span><span class="st"> </span><span class="va">$j</span><span class="st"> </span><span class="va">$new_path</span><span class="st"> &quot;</span></span>
<span id="cb16-27"><a href="#cb16-27" aria-hidden="true" tabindex="-1"></a>  <span class="cf">done</span><span class="kw">;</span>  </span>
<span id="cb16-28"><a href="#cb16-28" aria-hidden="true" tabindex="-1"></a><span class="cf">done</span></span>
<span id="cb16-29"><a href="#cb16-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-30"><a href="#cb16-30" aria-hidden="true" tabindex="-1"></a><span class="co"># How many files?</span></span>
<span id="cb16-31"><a href="#cb16-31" aria-hidden="true" tabindex="-1"></a><span class="fu">find</span> <span class="va">$segments_dir</span>/<span class="pp">*</span>/<span class="pp">*</span>.data.txt <span class="kw">|</span> <span class="fu">wc</span> <span class="at">-l</span></span>
<span id="cb16-32"><a href="#cb16-32" aria-hidden="true" tabindex="-1"></a><span class="co"># 8951</span></span>
<span id="cb16-33"><a href="#cb16-33" aria-hidden="true" tabindex="-1"></a><span class="fu">find</span> <span class="va">$in_dir</span>/<span class="pp">*</span>/<span class="pp">*</span>_1.data.txt <span class="kw">|</span> <span class="fu">wc</span> <span class="at">-l</span></span>
<span id="cb16-34"><a href="#cb16-34" aria-hidden="true" tabindex="-1"></a><span class="co"># 4341</span></span>
<span id="cb16-35"><a href="#cb16-35" aria-hidden="true" tabindex="-1"></a><span class="fu">find</span> <span class="va">$in_dir</span>/<span class="pp">*</span>/<span class="pp">*</span>_-1.data.txt <span class="kw">|</span> <span class="fu">wc</span> <span class="at">-l</span></span>
<span id="cb16-36"><a href="#cb16-36" aria-hidden="true" tabindex="-1"></a><span class="co"># 4610</span></span></code></pre></div>
</div>
<div id="run-analysis-pipeline-with-snakemake" class="section level2" number="2.4">
<h2 number="2.4"><span class="header-section-number">2.4</span> Run analysis pipeline with <code>snakemake</code></h2>
<div class="sourceCode" id="cb17"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="va">snmk_proj=</span><span class="st">&quot;introgression&quot;</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a><span class="ex">module</span> load singularity</span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> activate snakemake</span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a><span class="ex">snakemake</span> <span class="dt">\</span></span>
<span id="cb17-7"><a href="#cb17-7" aria-hidden="true" tabindex="-1"></a> <span class="at">--jobs</span> 5000 <span class="dt">\</span></span>
<span id="cb17-8"><a href="#cb17-8" aria-hidden="true" tabindex="-1"></a> <span class="at">--latency-wait</span> 100 <span class="dt">\</span></span>
<span id="cb17-9"><a href="#cb17-9" aria-hidden="true" tabindex="-1"></a> <span class="at">--cluster-config</span> code/snakemake/<span class="va">$snmk_proj</span>/config/cluster.json <span class="dt">\</span></span>
<span id="cb17-10"><a href="#cb17-10" aria-hidden="true" tabindex="-1"></a> <span class="at">--cluster</span> <span class="st">&#39;bsub -g /snakemake_bgenie -J {cluster.name} -n {cluster.n} -M {cluster.memory} -o {cluster.output} -e {cluster.error}&#39;</span> <span class="dt">\</span></span>
<span id="cb17-11"><a href="#cb17-11" aria-hidden="true" tabindex="-1"></a> <span class="at">--keep-going</span> <span class="dt">\</span></span>
<span id="cb17-12"><a href="#cb17-12" aria-hidden="true" tabindex="-1"></a> <span class="at">--rerun-incomplete</span> <span class="dt">\</span></span>
<span id="cb17-13"><a href="#cb17-13" aria-hidden="true" tabindex="-1"></a> <span class="at">--use-conda</span> <span class="dt">\</span></span>
<span id="cb17-14"><a href="#cb17-14" aria-hidden="true" tabindex="-1"></a> <span class="at">--use-singularity</span> <span class="dt">\</span></span>
<span id="cb17-15"><a href="#cb17-15" aria-hidden="true" tabindex="-1"></a> <span class="at">-s</span> code/snakemake/<span class="va">$snmk_proj</span>/Snakefile <span class="dt">\</span></span>
<span id="cb17-16"><a href="#cb17-16" aria-hidden="true" tabindex="-1"></a> <span class="at">-p</span></span></code></pre></div>
</div>
</div>
<div id="f-statistic-analysis" class="section level1" number="3">
<h1 number="3"><span class="header-section-number">3</span> <em>f</em> statistic analysis</h1>
<div id="read-in-data" class="section level2" number="3.1">
<h2 number="3.1"><span class="header-section-number">3.1</span> Read in data</h2>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>data_file <span class="ot">=</span> here<span class="sc">::</span><span class="fu">here</span>(<span class="st">&quot;data&quot;</span>, <span class="st">&quot;introgression&quot;</span>, <span class="st">&quot;20210315_f_stat_final.txt&quot;</span>)</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Read in data</span></span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>final_df <span class="ot">&lt;-</span> <span class="fu">read.table</span>(data_file,</span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a>                       <span class="at">header =</span> T,</span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a>                       <span class="at">sep =</span> <span class="st">&quot;</span><span class="sc">\t</span><span class="st">&quot;</span>,</span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a>                       <span class="at">as.is =</span> T)</span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-8"><a href="#cb18-8" aria-hidden="true" tabindex="-1"></a>final_df <span class="ot">&lt;-</span> final_df <span class="sc">%&gt;%</span> </span>
<span id="cb18-9"><a href="#cb18-9" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="fu">across</span>(P2,</span>
<span id="cb18-10"><a href="#cb18-10" aria-hidden="true" tabindex="-1"></a>                       <span class="sc">~</span><span class="fu">factor</span>(.x, <span class="at">levels =</span> fish_order))) <span class="sc">%&gt;%</span> </span>
<span id="cb18-11"><a href="#cb18-11" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">chr =</span> <span class="fu">factor</span>(chr, <span class="at">levels =</span> chr_order))</span>
<span id="cb18-12"><a href="#cb18-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-13"><a href="#cb18-13" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">kable</span>(<span class="fu">head</span>(final_df))</span></code></pre></div>
<table>
<thead>
<tr class="header">
<th align="left">P1</th>
<th align="left">P2</th>
<th align="left">P3</th>
<th align="left">chr</th>
<th align="right">d_stat</th>
<th align="right">z_score</th>
<th align="right">admix_f</th>
<th align="right">f_ci_lower</th>
<th align="right">f_ci_upper</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">javanicus</td>
<td align="left">HdrR</td>
<td align="left">MIKK</td>
<td align="left">all</td>
<td align="right">0.9072306</td>
<td align="right">324.06735</td>
<td align="right">0.7093524</td>
<td align="right">0.6943606</td>
<td align="right">0.7243441</td>
</tr>
<tr class="even">
<td align="left">javanicus</td>
<td align="left">HdrR</td>
<td align="left">MIKK</td>
<td align="left">1</td>
<td align="right">0.9422313</td>
<td align="right">162.46960</td>
<td align="right">0.8090713</td>
<td align="right">0.7776743</td>
<td align="right">0.8404682</td>
</tr>
<tr class="odd">
<td align="left">javanicus</td>
<td align="left">HdrR</td>
<td align="left">MIKK</td>
<td align="left">2</td>
<td align="right">0.9128441</td>
<td align="right">57.55129</td>
<td align="right">0.7668972</td>
<td align="right">0.6931469</td>
<td align="right">0.8406474</td>
</tr>
<tr class="even">
<td align="left">javanicus</td>
<td align="left">HdrR</td>
<td align="left">MIKK</td>
<td align="left">3</td>
<td align="right">0.8869725</td>
<td align="right">76.53532</td>
<td align="right">0.6217246</td>
<td align="right">0.5501172</td>
<td align="right">0.6933320</td>
</tr>
<tr class="odd">
<td align="left">javanicus</td>
<td align="left">HdrR</td>
<td align="left">MIKK</td>
<td align="left">4</td>
<td align="right">0.9318431</td>
<td align="right">72.86056</td>
<td align="right">0.7503846</td>
<td align="right">0.6643958</td>
<td align="right">0.8363734</td>
</tr>
<tr class="even">
<td align="left">javanicus</td>
<td align="left">HdrR</td>
<td align="left">MIKK</td>
<td align="left">5</td>
<td align="right">0.8879805</td>
<td align="right">69.67407</td>
<td align="right">0.6565419</td>
<td align="right">0.5865893</td>
<td align="right">0.7264945</td>
</tr>
</tbody>
</table>
</div>
<div id="create-df-with-mean-melastigma-and-javanicus" class="section level2" number="3.2">
<h2 number="3.2"><span class="header-section-number">3.2</span> Create DF with mean <code>melastigma</code> and <code>javanicus</code></h2>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>cor_df <span class="ot">&lt;-</span> final_df <span class="sc">%&gt;%</span> </span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>  <span class="co"># filter for when P1 is another Oryzias, and P2 </span></span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">filter</span>(P1 <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">&quot;javanicus&quot;</span>, <span class="st">&quot;melastigma&quot;</span>) <span class="sc">&amp;</span> P2 <span class="sc">!=</span> <span class="st">&quot;MIKK&quot;</span> <span class="sc">&amp;</span> P3 <span class="sc">==</span> <span class="st">&quot;MIKK&quot;</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a>  <span class="co"># pivot to put the admixture_f stat for melastigma and javanicus in the same row</span></span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a>  tidyr<span class="sc">::</span><span class="fu">pivot_wider</span>(<span class="at">id_cols =</span> <span class="fu">c</span>(<span class="st">&quot;P2&quot;</span>, <span class="st">&quot;chr&quot;</span>),</span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a>                     <span class="at">names_from =</span> P1,</span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true" tabindex="-1"></a>                     <span class="at">values_from =</span> <span class="fu">c</span>(admix_f, f_ci_lower, f_ci_upper))</span>
<span id="cb19-8"><a href="#cb19-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-9"><a href="#cb19-9" aria-hidden="true" tabindex="-1"></a>cor_df<span class="sc">$</span>chr <span class="ot">&lt;-</span> <span class="fu">as.character</span>(cor_df<span class="sc">$</span>chr)</span>
<span id="cb19-10"><a href="#cb19-10" aria-hidden="true" tabindex="-1"></a>cor_df<span class="sc">$</span>chr <span class="ot">&lt;-</span> <span class="fu">ifelse</span>(cor_df<span class="sc">$</span>chr <span class="sc">==</span> <span class="st">&quot;all&quot;</span>, <span class="st">&quot;genome-wide&quot;</span>, cor_df<span class="sc">$</span>chr)</span>
<span id="cb19-11"><a href="#cb19-11" aria-hidden="true" tabindex="-1"></a>chr_order_plot <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fu">seq</span>(<span class="dv">1</span>,<span class="dv">24</span>), <span class="st">&quot;genome-wide&quot;</span>)</span>
<span id="cb19-12"><a href="#cb19-12" aria-hidden="true" tabindex="-1"></a>cor_df<span class="sc">$</span>chr <span class="ot">&lt;-</span> <span class="fu">factor</span>(cor_df<span class="sc">$</span>chr, <span class="at">levels =</span> chr_order_plot)</span>
<span id="cb19-13"><a href="#cb19-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-14"><a href="#cb19-14" aria-hidden="true" tabindex="-1"></a>cor_df_means <span class="ot">&lt;-</span> cor_df <span class="sc">%&gt;%</span></span>
<span id="cb19-15"><a href="#cb19-15" aria-hidden="true" tabindex="-1"></a>  <span class="co"># apply across rows</span></span>
<span id="cb19-16"><a href="#cb19-16" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">rowwise</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb19-17"><a href="#cb19-17" aria-hidden="true" tabindex="-1"></a>  <span class="co"># get means for f and CIs</span></span>
<span id="cb19-18"><a href="#cb19-18" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">mean_f =</span> <span class="fu">mean</span>(<span class="fu">c</span>(admix_f_javanicus, admix_f_melastigma)),</span>
<span id="cb19-19"><a href="#cb19-19" aria-hidden="true" tabindex="-1"></a>                <span class="at">mean_ci_upper =</span> <span class="fu">mean</span>(<span class="fu">c</span>(f_ci_upper_javanicus, f_ci_upper_melastigma)),</span>
<span id="cb19-20"><a href="#cb19-20" aria-hidden="true" tabindex="-1"></a>                <span class="at">mean_ci_lower =</span> <span class="fu">mean</span>(<span class="fu">c</span>(f_ci_lower_javanicus, f_ci_lower_melastigma))) <span class="sc">%&gt;%</span> </span>
<span id="cb19-21"><a href="#cb19-21" aria-hidden="true" tabindex="-1"></a>  <span class="co"># set stats at a maximum of 1</span></span>
<span id="cb19-22"><a href="#cb19-22" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="fu">across</span>(<span class="fu">c</span>(<span class="st">&quot;mean_f&quot;</span>, <span class="st">&quot;mean_ci_upper&quot;</span>, <span class="st">&quot;mean_ci_lower&quot;</span>),</span>
<span id="cb19-23"><a href="#cb19-23" aria-hidden="true" tabindex="-1"></a>                       <span class="sc">~</span>dplyr<span class="sc">::</span><span class="fu">if_else</span>(.x <span class="sc">&gt;</span> <span class="dv">1</span>,</span>
<span id="cb19-24"><a href="#cb19-24" aria-hidden="true" tabindex="-1"></a>                                       <span class="dv">1</span>,</span>
<span id="cb19-25"><a href="#cb19-25" aria-hidden="true" tabindex="-1"></a>                                       .x)))</span>
<span id="cb19-26"><a href="#cb19-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-27"><a href="#cb19-27" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">kable</span>(<span class="fu">head</span>(cor_df_means))</span></code></pre></div>
<table style="width:100%;">
<colgroup>
<col width="2%" />
<col width="6%" />
<col width="10%" />
<col width="10%" />
<col width="11%" />
<col width="12%" />
<col width="11%" />
<col width="12%" />
<col width="5%" />
<col width="7%" />
<col width="7%" />
</colgroup>
<thead>
<tr class="header">
<th align="left">P2</th>
<th align="left">chr</th>
<th align="right">admix_f_javanicus</th>
<th align="right">admix_f_melastigma</th>
<th align="right">f_ci_lower_javanicus</th>
<th align="right">f_ci_lower_melastigma</th>
<th align="right">f_ci_upper_javanicus</th>
<th align="right">f_ci_upper_melastigma</th>
<th align="right">mean_f</th>
<th align="right">mean_ci_upper</th>
<th align="right">mean_ci_lower</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">HdrR</td>
<td align="left">genome-wide</td>
<td align="right">0.7093524</td>
<td align="right">0.6860648</td>
<td align="right">0.6943606</td>
<td align="right">0.6712495</td>
<td align="right">0.7243441</td>
<td align="right">0.7008800</td>
<td align="right">0.6977086</td>
<td align="right">0.7126121</td>
<td align="right">0.6828051</td>
</tr>
<tr class="even">
<td align="left">HdrR</td>
<td align="left">1</td>
<td align="right">0.8090713</td>
<td align="right">0.7831495</td>
<td align="right">0.7776743</td>
<td align="right">0.7466954</td>
<td align="right">0.8404682</td>
<td align="right">0.8196036</td>
<td align="right">0.7961104</td>
<td align="right">0.8300359</td>
<td align="right">0.7621849</td>
</tr>
<tr class="odd">
<td align="left">HdrR</td>
<td align="left">2</td>
<td align="right">0.7668972</td>
<td align="right">0.7621796</td>
<td align="right">0.6931469</td>
<td align="right">0.6999464</td>
<td align="right">0.8406474</td>
<td align="right">0.8244128</td>
<td align="right">0.7645384</td>
<td align="right">0.8325301</td>
<td align="right">0.6965466</td>
</tr>
<tr class="even">
<td align="left">HdrR</td>
<td align="left">3</td>
<td align="right">0.6217246</td>
<td align="right">0.5968322</td>
<td align="right">0.5501172</td>
<td align="right">0.5217981</td>
<td align="right">0.6933320</td>
<td align="right">0.6718664</td>
<td align="right">0.6092784</td>
<td align="right">0.6825992</td>
<td align="right">0.5359576</td>
</tr>
<tr class="odd">
<td align="left">HdrR</td>
<td align="left">4</td>
<td align="right">0.7503846</td>
<td align="right">0.7399951</td>
<td align="right">0.6643958</td>
<td align="right">0.6811117</td>
<td align="right">0.8363734</td>
<td align="right">0.7988785</td>
<td align="right">0.7451899</td>
<td align="right">0.8176260</td>
<td align="right">0.6727538</td>
</tr>
<tr class="even">
<td align="left">HdrR</td>
<td align="left">5</td>
<td align="right">0.6565419</td>
<td align="right">0.6241430</td>
<td align="right">0.5865893</td>
<td align="right">0.5565435</td>
<td align="right">0.7264945</td>
<td align="right">0.6917426</td>
<td align="right">0.6403425</td>
<td align="right">0.7091186</td>
<td align="right">0.5715664</td>
</tr>
</tbody>
</table>
</div>
<div id="plot" class="section level2" number="3.3">
<h2 number="3.3"><span class="header-section-number">3.3</span> Plot</h2>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>fstat_plot <span class="ot">=</span> cor_df_means <span class="sc">%&gt;%</span> </span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(P2, mean_f, <span class="at">fill =</span> P2)) <span class="sc">+</span></span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_col</span>() <span class="sc">+</span></span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_errorbar</span>(<span class="fu">aes</span>(<span class="at">ymin =</span> mean_ci_lower,</span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a>                      <span class="at">ymax =</span> mean_ci_upper),</span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a>                  <span class="at">position =</span> <span class="fu">position_dodge</span>(<span class="fl">0.9</span>),</span>
<span id="cb20-7"><a href="#cb20-7" aria-hidden="true" tabindex="-1"></a>                  <span class="at">width =</span> <span class="fl">0.25</span>) <span class="sc">+</span>  </span>
<span id="cb20-8"><a href="#cb20-8" aria-hidden="true" tabindex="-1"></a>    <span class="fu">guides</span>(<span class="at">fill =</span> F) <span class="sc">+</span></span>
<span id="cb20-9"><a href="#cb20-9" aria-hidden="true" tabindex="-1"></a>    <span class="fu">facet_wrap</span>(<span class="sc">~</span>chr) <span class="sc">+</span></span>
<span id="cb20-10"><a href="#cb20-10" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ylim</span>(<span class="dv">0</span>,<span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb20-11"><a href="#cb20-11" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ylab</span>(<span class="fu">expression</span>(<span class="fu">paste</span>(<span class="st">&quot;Mean &quot;</span>, <span class="fu">italic</span>(<span class="st">&quot;f&quot;</span>), <span class="st">&quot; statistic&quot;</span>))) <span class="sc">+</span></span>
<span id="cb20-12"><a href="#cb20-12" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_cowplot</span>(<span class="at">font_size =</span> <span class="dv">8</span>) <span class="sc">+</span></span>
<span id="cb20-13"><a href="#cb20-13" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> pal_abba)</span>
<span id="cb20-14"><a href="#cb20-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-15"><a href="#cb20-15" aria-hidden="true" tabindex="-1"></a>fstat_plot</span></code></pre></div>
<p><img src="20210318_introgression_files/figure-html/unnamed-chunk-21-1.svg" width="672" /></p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>out_file <span class="ot">=</span> here<span class="sc">::</span><span class="fu">here</span>(plots_dir, <span class="st">&quot;20210315_f_stat&quot;</span>)</span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a><span class="co"># PNG</span></span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="at">filename =</span> <span class="fu">paste</span>(out_file, <span class="st">&quot;.png&quot;</span>, <span class="at">sep =</span> <span class="st">&quot;&quot;</span>),</span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a>       <span class="at">device =</span> <span class="st">&quot;png&quot;</span>,</span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true" tabindex="-1"></a>       <span class="at">width =</span> <span class="fl">24.75</span>,</span>
<span id="cb21-7"><a href="#cb21-7" aria-hidden="true" tabindex="-1"></a>       <span class="at">height =</span> <span class="fl">19.5</span>,</span>
<span id="cb21-8"><a href="#cb21-8" aria-hidden="true" tabindex="-1"></a>       <span class="at">units =</span> <span class="st">&quot;cm&quot;</span>,</span>
<span id="cb21-9"><a href="#cb21-9" aria-hidden="true" tabindex="-1"></a>       <span class="at">dpi =</span> <span class="dv">500</span>)</span>
<span id="cb21-10"><a href="#cb21-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-11"><a href="#cb21-11" aria-hidden="true" tabindex="-1"></a><span class="co"># SVG</span></span>
<span id="cb21-12"><a href="#cb21-12" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="at">filename =</span> <span class="fu">paste</span>(out_file, <span class="st">&quot;.svg&quot;</span>, <span class="at">sep =</span> <span class="st">&quot;&quot;</span>),</span>
<span id="cb21-13"><a href="#cb21-13" aria-hidden="true" tabindex="-1"></a>       <span class="at">device =</span> <span class="st">&quot;svg&quot;</span>,</span>
<span id="cb21-14"><a href="#cb21-14" aria-hidden="true" tabindex="-1"></a>       <span class="at">width =</span> <span class="fl">24.75</span>,</span>
<span id="cb21-15"><a href="#cb21-15" aria-hidden="true" tabindex="-1"></a>       <span class="at">height =</span> <span class="fl">19.5</span>,</span>
<span id="cb21-16"><a href="#cb21-16" aria-hidden="true" tabindex="-1"></a>       <span class="at">units =</span> <span class="st">&quot;cm&quot;</span>)</span></code></pre></div>
</div>
</div>
<div id="sliding-windows-abba-baba" class="section level1" number="4">
<h1 number="4"><span class="header-section-number">4</span> Sliding windows ABBA BABA</h1>
<div id="read-in-data-1" class="section level2" number="4.1">
<h2 number="4.1"><span class="header-section-number">4.1</span> Read in data</h2>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>in_file <span class="ot">=</span> here<span class="sc">::</span><span class="fu">here</span>(<span class="st">&quot;data&quot;</span>, <span class="st">&quot;introgression&quot;</span>, <span class="st">&quot;abba_sliding_final&quot;</span>, <span class="st">&quot;50000_100.txt&quot;</span>)</span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Read in data</span></span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a>df <span class="ot">=</span> readr<span class="sc">::</span><span class="fu">read_csv</span>(in_file) <span class="sc">%&gt;%</span> </span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">arrange</span>(p1, p2, scaffold, start)</span>
<span id="cb22-5"><a href="#cb22-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-6"><a href="#cb22-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Convert fd to 0 if D &lt; 0</span></span>
<span id="cb22-7"><a href="#cb22-7" aria-hidden="true" tabindex="-1"></a>df<span class="sc">$</span>fd <span class="ot">=</span> <span class="fu">ifelse</span>(df<span class="sc">$</span>D <span class="sc">&lt;</span> <span class="dv">0</span>,</span>
<span id="cb22-8"><a href="#cb22-8" aria-hidden="true" tabindex="-1"></a>               <span class="dv">0</span>,</span>
<span id="cb22-9"><a href="#cb22-9" aria-hidden="true" tabindex="-1"></a>               df<span class="sc">$</span>fd)</span>
<span id="cb22-10"><a href="#cb22-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-11"><a href="#cb22-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Change names</span></span>
<span id="cb22-12"><a href="#cb22-12" aria-hidden="true" tabindex="-1"></a>df <span class="ot">=</span> df <span class="sc">%&gt;%</span> </span>
<span id="cb22-13"><a href="#cb22-13" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">p2 =</span> <span class="fu">recode</span>(df<span class="sc">$</span>p2, <span class="at">hdrr =</span> <span class="st">&quot;HdrR&quot;</span>, <span class="at">hni =</span> <span class="st">&quot;HNI&quot;</span>, <span class="at">hsok =</span> <span class="st">&quot;HSOK&quot;</span>))</span></code></pre></div>
</div>
<div id="plot-1" class="section level2" number="4.2">
<h2 number="4.2"><span class="header-section-number">4.2</span> Plot</h2>
<div id="standard" class="section level3" number="4.2.1">
<h3 number="4.2.1"><span class="header-section-number">4.2.1</span> Standard</h3>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>df <span class="sc">%&gt;%</span> </span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">filter</span>(p1 <span class="sc">==</span> <span class="st">&quot;melastigma&quot;</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_line</span>(<span class="fu">aes</span>(mid, fd, <span class="at">colour =</span> p2)) <span class="sc">+</span></span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">facet_wrap</span>(<span class="sc">~</span>scaffold, <span class="at">nrow =</span> <span class="dv">24</span>, <span class="at">ncol =</span> <span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb23-6"><a href="#cb23-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_colour_manual</span>(<span class="at">values =</span> pal_abba) <span class="sc">+</span></span>
<span id="cb23-7"><a href="#cb23-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_bw</span>(<span class="at">base_size =</span> <span class="dv">10</span>) <span class="sc">+</span></span>
<span id="cb23-8"><a href="#cb23-8" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_x_continuous</span>(<span class="at">breaks =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">5000000</span>, <span class="dv">10000000</span>, <span class="dv">15000000</span>, <span class="dv">20000000</span>, <span class="dv">25000000</span>, <span class="dv">30000000</span>, <span class="dv">35000000</span>),</span>
<span id="cb23-9"><a href="#cb23-9" aria-hidden="true" tabindex="-1"></a>                       <span class="at">labels =</span> scales<span class="sc">::</span>comma) <span class="sc">+</span></span>
<span id="cb23-10"><a href="#cb23-10" aria-hidden="true" tabindex="-1"></a>    <span class="fu">xlab</span>(<span class="st">&quot;Base position&quot;</span>) <span class="sc">+</span></span>
<span id="cb23-11"><a href="#cb23-11" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ylab</span>(<span class="fu">bquote</span>(<span class="fu">italic</span>(f[d]))) <span class="sc">+</span></span>
<span id="cb23-12"><a href="#cb23-12" aria-hidden="true" tabindex="-1"></a>    <span class="fu">labs</span>(<span class="at">colour =</span> <span class="st">&quot;P2&quot;</span>)</span></code></pre></div>
<p><img src="20210318_introgression_files/figure-html/unnamed-chunk-24-1.svg" width="960" /></p>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>out_file <span class="ot">=</span> here<span class="sc">::</span><span class="fu">here</span>(plots_dir, <span class="st">&quot;20210317_abba_sliding.png&quot;</span>)</span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="at">filename =</span> out_file,</span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a>       <span class="at">device =</span> <span class="st">&quot;png&quot;</span>,</span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a>       <span class="at">width =</span> <span class="fl">24.75</span>,</span>
<span id="cb24-6"><a href="#cb24-6" aria-hidden="true" tabindex="-1"></a>       <span class="at">height =</span> <span class="dv">50</span>,</span>
<span id="cb24-7"><a href="#cb24-7" aria-hidden="true" tabindex="-1"></a>       <span class="at">units =</span> <span class="st">&quot;cm&quot;</span>,</span>
<span id="cb24-8"><a href="#cb24-8" aria-hidden="true" tabindex="-1"></a>       <span class="at">dpi =</span> <span class="dv">300</span>)</span></code></pre></div>
</div>
<div id="karyoplot" class="section level3" number="4.2.2">
<h3 number="4.2.2"><span class="header-section-number">4.2.2</span> Karyoplot</h3>
<div id="make-custom-chromosome-scaffold" class="section level4" number="4.2.2.1">
<h4 number="4.2.2.1"><span class="header-section-number">4.2.2.1</span> Make custom chromosome scaffold</h4>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Get chromosome lengths</span></span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>med_chr_lens <span class="ot">=</span> <span class="fu">read.table</span>(<span class="fu">here</span>(<span class="st">&quot;data&quot;</span>, <span class="st">&quot;Oryzias_latipes.ASM223467v1.dna.toplevel.fa_chr_counts.txt&quot;</span>),</span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a>                          <span class="at">col.names =</span> <span class="fu">c</span>(<span class="st">&quot;chr&quot;</span>, <span class="st">&quot;end&quot;</span>))</span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Add start</span></span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a>med_chr_lens<span class="sc">$</span>start <span class="ot">=</span> <span class="dv">1</span></span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Reorder</span></span>
<span id="cb25-7"><a href="#cb25-7" aria-hidden="true" tabindex="-1"></a>med_chr_lens <span class="ot">=</span> med_chr_lens <span class="sc">%&gt;%</span> </span>
<span id="cb25-8"><a href="#cb25-8" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(chr, start, end)</span>
<span id="cb25-9"><a href="#cb25-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Create custom genome</span></span>
<span id="cb25-10"><a href="#cb25-10" aria-hidden="true" tabindex="-1"></a>med_genome <span class="ot">=</span> regioneR<span class="sc">::</span><span class="fu">toGRanges</span>(med_chr_lens)</span></code></pre></div>
</div>
<div id="process-abba-sliding-windows-data" class="section level4" number="4.2.2.2">
<h4 number="4.2.2.2"><span class="header-section-number">4.2.2.2</span> Process ABBA sliding windows data</h4>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>in_file <span class="ot">=</span> here<span class="sc">::</span><span class="fu">here</span>(<span class="st">&quot;data&quot;</span>, <span class="st">&quot;introgression&quot;</span>, <span class="st">&quot;abba_sliding_final&quot;</span>, <span class="st">&quot;50000_100.txt&quot;</span>)</span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Read in data</span></span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a>df <span class="ot">=</span> readr<span class="sc">::</span><span class="fu">read_csv</span>(in_file) <span class="sc">%&gt;%</span> </span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">arrange</span>(p1, p2, scaffold, start)</span>
<span id="cb26-5"><a href="#cb26-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-6"><a href="#cb26-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Convert fd to 0 if D &lt; 0</span></span>
<span id="cb26-7"><a href="#cb26-7" aria-hidden="true" tabindex="-1"></a>df<span class="sc">$</span>fd <span class="ot">=</span> <span class="fu">ifelse</span>(df<span class="sc">$</span>D <span class="sc">&lt;</span> <span class="dv">0</span>,</span>
<span id="cb26-8"><a href="#cb26-8" aria-hidden="true" tabindex="-1"></a>               <span class="dv">0</span>,</span>
<span id="cb26-9"><a href="#cb26-9" aria-hidden="true" tabindex="-1"></a>               df<span class="sc">$</span>fd)</span>
<span id="cb26-10"><a href="#cb26-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-11"><a href="#cb26-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Change names</span></span>
<span id="cb26-12"><a href="#cb26-12" aria-hidden="true" tabindex="-1"></a>df <span class="ot">=</span> df <span class="sc">%&gt;%</span> </span>
<span id="cb26-13"><a href="#cb26-13" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">p2 =</span> <span class="fu">recode</span>(df<span class="sc">$</span>p2, <span class="at">hdrr =</span> <span class="st">&quot;HdrR&quot;</span>, <span class="at">hni =</span> <span class="st">&quot;HNI&quot;</span>, <span class="at">hsok =</span> <span class="st">&quot;HSOK&quot;</span>))</span>
<span id="cb26-14"><a href="#cb26-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-15"><a href="#cb26-15" aria-hidden="true" tabindex="-1"></a><span class="co"># Get df with mean of melastigma/javanicus</span></span>
<span id="cb26-16"><a href="#cb26-16" aria-hidden="true" tabindex="-1"></a>df_kp <span class="ot">=</span> df <span class="sc">%&gt;%</span> </span>
<span id="cb26-17"><a href="#cb26-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_wider</span>(<span class="at">id_cols =</span> <span class="fu">c</span>(scaffold, start, end, mid, p2), <span class="at">names_from =</span> p1, <span class="at">values_from =</span> fd) <span class="sc">%&gt;%</span></span>
<span id="cb26-18"><a href="#cb26-18" aria-hidden="true" tabindex="-1"></a>  <span class="co"># get mean of melastigma/javanicus</span></span>
<span id="cb26-19"><a href="#cb26-19" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">mean_fd =</span> <span class="fu">rowMeans</span>(dplyr<span class="sc">::</span><span class="fu">select</span>(., melastigma, javanicus), <span class="at">na.rm =</span> T)) <span class="sc">%&gt;%</span> </span>
<span id="cb26-20"><a href="#cb26-20" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">arrange</span>(p2, scaffold, start)</span>
<span id="cb26-21"><a href="#cb26-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-22"><a href="#cb26-22" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">kable</span>(<span class="fu">head</span>(df_kp))</span></code></pre></div>
<table>
<thead>
<tr class="header">
<th align="right">scaffold</th>
<th align="right">start</th>
<th align="right">end</th>
<th align="right">mid</th>
<th align="left">p2</th>
<th align="right">javanicus</th>
<th align="right">melastigma</th>
<th align="right">mean_fd</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="right">1</td>
<td align="right">650001</td>
<td align="right">700000</td>
<td align="right">666130</td>
<td align="left">HdrR</td>
<td align="right">NA</td>
<td align="right">0.1231</td>
<td align="right">0.12310</td>
</tr>
<tr class="even">
<td align="right">1</td>
<td align="right">700001</td>
<td align="right">750000</td>
<td align="right">721651</td>
<td align="left">HdrR</td>
<td align="right">NA</td>
<td align="right">0.2357</td>
<td align="right">0.23570</td>
</tr>
<tr class="odd">
<td align="right">1</td>
<td align="right">750001</td>
<td align="right">800000</td>
<td align="right">774678</td>
<td align="left">HdrR</td>
<td align="right">NA</td>
<td align="right">0.1636</td>
<td align="right">0.16360</td>
</tr>
<tr class="even">
<td align="right">1</td>
<td align="right">950001</td>
<td align="right">1000000</td>
<td align="right">974734</td>
<td align="left">HdrR</td>
<td align="right">NA</td>
<td align="right">0.2203</td>
<td align="right">0.22030</td>
</tr>
<tr class="odd">
<td align="right">1</td>
<td align="right">1000001</td>
<td align="right">1050000</td>
<td align="right">1028780</td>
<td align="left">HdrR</td>
<td align="right">NA</td>
<td align="right">0.2935</td>
<td align="right">0.29350</td>
</tr>
<tr class="even">
<td align="right">1</td>
<td align="right">1050001</td>
<td align="right">1100000</td>
<td align="right">1075868</td>
<td align="left">HdrR</td>
<td align="right">0.4066</td>
<td align="right">0.2985</td>
<td align="right">0.35255</td>
</tr>
</tbody>
</table>
</div>
<div id="read-in-snp-density-data" class="section level4" number="4.2.2.3">
<h4 number="4.2.2.3"><span class="header-section-number">4.2.2.3</span> Read in SNP density data</h4>
<div id="hni-and-hsok" class="section level5" number="4.2.2.3.1">
<h5 number="4.2.2.3.1"><span class="header-section-number">4.2.2.3.1</span> HNI and HSOK</h5>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>in_file <span class="ot">=</span> here<span class="sc">::</span><span class="fu">here</span>(<span class="st">&quot;data/introgression/hni_hsok.txt.gz&quot;</span>)</span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Read in file on local</span></span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a>ol_ranges_df <span class="ot">=</span> <span class="fu">read.table</span>(in_file,</span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a>                          <span class="at">header =</span> T, </span>
<span id="cb27-5"><a href="#cb27-5" aria-hidden="true" tabindex="-1"></a>                          <span class="at">sep =</span> <span class="st">&quot;</span><span class="sc">\t</span><span class="st">&quot;</span>, </span>
<span id="cb27-6"><a href="#cb27-6" aria-hidden="true" tabindex="-1"></a>                          <span class="at">as.is =</span> T)</span>
<span id="cb27-7"><a href="#cb27-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-8"><a href="#cb27-8" aria-hidden="true" tabindex="-1"></a>ol_ranges_df_long <span class="ot">=</span> ol_ranges_df <span class="sc">%&gt;%</span> </span>
<span id="cb27-9"><a href="#cb27-9" aria-hidden="true" tabindex="-1"></a>  tidyr<span class="sc">::</span><span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="fu">c</span>(hni, hsok), <span class="at">names_to =</span> <span class="st">&quot;line&quot;</span>, <span class="at">values_to =</span> <span class="st">&quot;present&quot;</span>)</span>
<span id="cb27-10"><a href="#cb27-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-11"><a href="#cb27-11" aria-hidden="true" tabindex="-1"></a>ol_ranges_list <span class="ot">=</span> <span class="fu">split</span>(ol_ranges_df_long, <span class="at">f =</span> ol_ranges_df_long<span class="sc">$</span>line)</span>
<span id="cb27-12"><a href="#cb27-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-13"><a href="#cb27-13" aria-hidden="true" tabindex="-1"></a>ol_ranges_list <span class="ot">=</span> <span class="fu">lapply</span>(ol_ranges_list, <span class="cf">function</span>(x){</span>
<span id="cb27-14"><a href="#cb27-14" aria-hidden="true" tabindex="-1"></a>  <span class="co"># remove NAs</span></span>
<span id="cb27-15"><a href="#cb27-15" aria-hidden="true" tabindex="-1"></a>  df <span class="ot">=</span> x <span class="sc">%&gt;%</span> </span>
<span id="cb27-16"><a href="#cb27-16" aria-hidden="true" tabindex="-1"></a>    tidyr<span class="sc">::</span><span class="fu">drop_na</span>(present)</span>
<span id="cb27-17"><a href="#cb27-17" aria-hidden="true" tabindex="-1"></a>  <span class="co"># convert to GRanges object</span></span>
<span id="cb27-18"><a href="#cb27-18" aria-hidden="true" tabindex="-1"></a>  ol_ranges <span class="ot">=</span> GenomicRanges<span class="sc">::</span><span class="fu">makeGRangesFromDataFrame</span>(df,</span>
<span id="cb27-19"><a href="#cb27-19" aria-hidden="true" tabindex="-1"></a>                                                      <span class="at">ignore.strand =</span> T,</span>
<span id="cb27-20"><a href="#cb27-20" aria-hidden="true" tabindex="-1"></a>                                                      <span class="at">seqnames.field =</span> <span class="st">&quot;chr&quot;</span>,</span>
<span id="cb27-21"><a href="#cb27-21" aria-hidden="true" tabindex="-1"></a>                                                      <span class="at">start.field =</span> <span class="st">&quot;pos&quot;</span>,</span>
<span id="cb27-22"><a href="#cb27-22" aria-hidden="true" tabindex="-1"></a>                                                      <span class="at">end.field =</span> <span class="st">&quot;pos&quot;</span>)</span>
<span id="cb27-23"><a href="#cb27-23" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(ol_ranges)</span>
<span id="cb27-24"><a href="#cb27-24" aria-hidden="true" tabindex="-1"></a>})</span></code></pre></div>
</div>
<div id="mikk" class="section level5" number="4.2.2.3.2">
<h5 number="4.2.2.3.2"><span class="header-section-number">4.2.2.3.2</span> MIKK</h5>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>in_file <span class="ot">=</span> here<span class="sc">::</span><span class="fu">here</span>(<span class="st">&quot;data/introgression/mikk.txt.gz&quot;</span>)</span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Read in file on local</span></span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a>mikk_ranges_df <span class="ot">=</span> <span class="fu">read.table</span>(in_file,</span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a>                            <span class="at">col.names =</span> <span class="fu">c</span>(<span class="st">&quot;chr&quot;</span>, <span class="st">&quot;pos&quot;</span>), </span>
<span id="cb28-5"><a href="#cb28-5" aria-hidden="true" tabindex="-1"></a>                            <span class="at">sep =</span> <span class="st">&quot;</span><span class="sc">\t</span><span class="st">&quot;</span>, </span>
<span id="cb28-6"><a href="#cb28-6" aria-hidden="true" tabindex="-1"></a>                            <span class="at">as.is =</span> T)</span>
<span id="cb28-7"><a href="#cb28-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-8"><a href="#cb28-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Convert to GRanges object</span></span>
<span id="cb28-9"><a href="#cb28-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-10"><a href="#cb28-10" aria-hidden="true" tabindex="-1"></a>mikk_ranges <span class="ot">=</span> GenomicRanges<span class="sc">::</span><span class="fu">makeGRangesFromDataFrame</span>(mikk_ranges_df,</span>
<span id="cb28-11"><a href="#cb28-11" aria-hidden="true" tabindex="-1"></a>                                                      <span class="at">ignore.strand =</span> T,</span>
<span id="cb28-12"><a href="#cb28-12" aria-hidden="true" tabindex="-1"></a>                                                      <span class="at">seqnames.field =</span> <span class="st">&quot;chr&quot;</span>,</span>
<span id="cb28-13"><a href="#cb28-13" aria-hidden="true" tabindex="-1"></a>                                                      <span class="at">start.field =</span> <span class="st">&quot;pos&quot;</span>,</span>
<span id="cb28-14"><a href="#cb28-14" aria-hidden="true" tabindex="-1"></a>                                                      <span class="at">end.field =</span> <span class="st">&quot;pos&quot;</span>)</span></code></pre></div>
</div>
</div>
<div id="get-exon-density" class="section level4" number="4.2.2.4">
<h4 number="4.2.2.4"><span class="header-section-number">4.2.2.4</span> Get exon density</h4>
<div class="sourceCode" id="cb29"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Get list of exons from biomaRt</span></span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a><span class="do">## Select dataset</span></span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a>olat_mart <span class="ot">=</span> biomaRt<span class="sc">::</span><span class="fu">useEnsembl</span>(<span class="at">biomart =</span> <span class="st">&quot;ensembl&quot;</span>, <span class="at">dataset =</span> <span class="st">&quot;olatipes_gene_ensembl&quot;</span>)</span>
<span id="cb29-5"><a href="#cb29-5" aria-hidden="true" tabindex="-1"></a><span class="do">## Get attributes of interest (exon ID, chr, start, end)</span></span>
<span id="cb29-6"><a href="#cb29-6" aria-hidden="true" tabindex="-1"></a>exons <span class="ot">&lt;-</span> biomaRt<span class="sc">::</span><span class="fu">getBM</span>(<span class="at">attributes =</span> <span class="fu">c</span>(<span class="st">&quot;chromosome_name&quot;</span>, <span class="st">&quot;ensembl_gene_id&quot;</span>, <span class="st">&quot;ensembl_transcript_id&quot;</span>, <span class="st">&quot;transcript_start&quot;</span>, <span class="st">&quot;transcript_end&quot;</span>, <span class="st">&quot;transcript_length&quot;</span>, <span class="st">&quot;ensembl_exon_id&quot;</span>, <span class="st">&quot;rank&quot;</span>, <span class="st">&quot;strand&quot;</span>, <span class="st">&quot;exon_chrom_start&quot;</span>, <span class="st">&quot;exon_chrom_end&quot;</span>, <span class="st">&quot;cds_start&quot;</span>, <span class="st">&quot;cds_end&quot;</span>),</span>
<span id="cb29-7"><a href="#cb29-7" aria-hidden="true" tabindex="-1"></a>               <span class="at">mart =</span> olat_mart)</span>
<span id="cb29-8"><a href="#cb29-8" aria-hidden="true" tabindex="-1"></a><span class="do">## Convert exons to GRanges</span></span>
<span id="cb29-9"><a href="#cb29-9" aria-hidden="true" tabindex="-1"></a>ex_ranges <span class="ot">=</span> GenomicRanges<span class="sc">::</span><span class="fu">makeGRangesFromDataFrame</span>(exons,</span>
<span id="cb29-10"><a href="#cb29-10" aria-hidden="true" tabindex="-1"></a>                                                    <span class="at">ignore.strand =</span> T,</span>
<span id="cb29-11"><a href="#cb29-11" aria-hidden="true" tabindex="-1"></a>                                                    <span class="at">seqnames.field =</span> <span class="st">&quot;chromosome_name&quot;</span>,</span>
<span id="cb29-12"><a href="#cb29-12" aria-hidden="true" tabindex="-1"></a>                                                    <span class="at">start.field =</span> <span class="st">&quot;exon_chrom_start&quot;</span>,</span>
<span id="cb29-13"><a href="#cb29-13" aria-hidden="true" tabindex="-1"></a>                                                    <span class="at">end.field =</span> <span class="st">&quot;exon_chrom_end&quot;</span>)</span></code></pre></div>
</div>
<div id="all-chromosomes" class="section level4" number="4.2.2.5">
<h4 number="4.2.2.5"><span class="header-section-number">4.2.2.5</span> All chromosomes</h4>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>file_out <span class="ot">=</span> <span class="fu">file.path</span>(plots_dir, <span class="st">&quot;20210318_fd_with_density_all.png&quot;</span>)</span></code></pre></div>
<div class="sourceCode" id="cb31"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Save</span></span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a><span class="fu">png</span>(<span class="at">file=</span>file_out,</span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">width=</span><span class="dv">8500</span>,</span>
<span id="cb31-4"><a href="#cb31-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">height=</span><span class="dv">13500</span>,</span>
<span id="cb31-5"><a href="#cb31-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">units =</span> <span class="st">&quot;px&quot;</span>,</span>
<span id="cb31-6"><a href="#cb31-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">res =</span> <span class="dv">400</span>)</span>
<span id="cb31-7"><a href="#cb31-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot </span></span>
<span id="cb31-8"><a href="#cb31-8" aria-hidden="true" tabindex="-1"></a>kp <span class="ot">=</span> <span class="fu">plotKaryotype</span>(med_genome)</span>
<span id="cb31-9"><a href="#cb31-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Add base numbers </span></span>
<span id="cb31-10"><a href="#cb31-10" aria-hidden="true" tabindex="-1"></a>karyoploteR<span class="sc">::</span><span class="fu">kpAddBaseNumbers</span>(kp, <span class="at">tick.dist =</span> <span class="dv">5000000</span>, <span class="at">cex =</span> <span class="fl">0.3</span>)</span>
<span id="cb31-11"><a href="#cb31-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Add data backgrounds</span></span>
<span id="cb31-12"><a href="#cb31-12" aria-hidden="true" tabindex="-1"></a>karyoploteR<span class="sc">::</span><span class="fu">kpDataBackground</span>(kp, <span class="at">r0=</span><span class="dv">0</span>, <span class="at">r1 =</span> <span class="dv">1</span>, <span class="at">color =</span> <span class="st">&quot;white&quot;</span>)</span>
<span id="cb31-13"><a href="#cb31-13" aria-hidden="true" tabindex="-1"></a><span class="co"># Add axis label</span></span>
<span id="cb31-14"><a href="#cb31-14" aria-hidden="true" tabindex="-1"></a><span class="fu">kpAxis</span>(kp, <span class="at">r0=</span><span class="fl">0.3</span>, <span class="at">r1 =</span> <span class="dv">1</span>, <span class="at">cex =</span> <span class="fl">0.4</span>)</span>
<span id="cb31-15"><a href="#cb31-15" aria-hidden="true" tabindex="-1"></a><span class="co"># Add fd data</span></span>
<span id="cb31-16"><a href="#cb31-16" aria-hidden="true" tabindex="-1"></a>karyoploteR<span class="sc">::</span><span class="fu">kpLines</span>(kp,</span>
<span id="cb31-17"><a href="#cb31-17" aria-hidden="true" tabindex="-1"></a>                     <span class="at">chr =</span> df_kp<span class="sc">$</span>scaffold[df_kp<span class="sc">$</span>p2 <span class="sc">==</span> <span class="st">&quot;HNI&quot;</span>],</span>
<span id="cb31-18"><a href="#cb31-18" aria-hidden="true" tabindex="-1"></a>                     <span class="at">x =</span> df_kp<span class="sc">$</span>mid[df_kp<span class="sc">$</span>p2 <span class="sc">==</span> <span class="st">&quot;HNI&quot;</span>],</span>
<span id="cb31-19"><a href="#cb31-19" aria-hidden="true" tabindex="-1"></a>                     <span class="at">y =</span> df_kp<span class="sc">$</span>mean_fd[df_kp<span class="sc">$</span>p2 <span class="sc">==</span> <span class="st">&quot;HNI&quot;</span>],</span>
<span id="cb31-20"><a href="#cb31-20" aria-hidden="true" tabindex="-1"></a>                     <span class="at">col =</span> <span class="st">&quot;#F6673A&quot;</span>,</span>
<span id="cb31-21"><a href="#cb31-21" aria-hidden="true" tabindex="-1"></a>                     <span class="at">r0=</span><span class="fl">0.3</span>, <span class="at">r1 =</span> <span class="dv">1</span>)</span>
<span id="cb31-22"><a href="#cb31-22" aria-hidden="true" tabindex="-1"></a>karyoploteR<span class="sc">::</span><span class="fu">kpLines</span>(kp,</span>
<span id="cb31-23"><a href="#cb31-23" aria-hidden="true" tabindex="-1"></a>                     <span class="at">chr =</span> df_kp<span class="sc">$</span>scaffold[df_kp<span class="sc">$</span>p2 <span class="sc">==</span> <span class="st">&quot;HdrR&quot;</span>],</span>
<span id="cb31-24"><a href="#cb31-24" aria-hidden="true" tabindex="-1"></a>                     <span class="at">x =</span> df_kp<span class="sc">$</span>mid[df_kp<span class="sc">$</span>p2 <span class="sc">==</span> <span class="st">&quot;HdrR&quot;</span>],</span>
<span id="cb31-25"><a href="#cb31-25" aria-hidden="true" tabindex="-1"></a>                     <span class="at">y =</span> df_kp<span class="sc">$</span>mean_fd[df_kp<span class="sc">$</span>p2 <span class="sc">==</span> <span class="st">&quot;HdrR&quot;</span>],</span>
<span id="cb31-26"><a href="#cb31-26" aria-hidden="true" tabindex="-1"></a>                     <span class="at">col =</span> <span class="st">&quot;#F3B61F&quot;</span>,</span>
<span id="cb31-27"><a href="#cb31-27" aria-hidden="true" tabindex="-1"></a>                     <span class="at">r0=</span><span class="fl">0.3</span>, <span class="at">r1 =</span> <span class="dv">1</span>)</span>
<span id="cb31-28"><a href="#cb31-28" aria-hidden="true" tabindex="-1"></a>karyoploteR<span class="sc">::</span><span class="fu">kpLines</span>(kp,</span>
<span id="cb31-29"><a href="#cb31-29" aria-hidden="true" tabindex="-1"></a>                     <span class="at">chr =</span> df_kp<span class="sc">$</span>scaffold[df_kp<span class="sc">$</span>p2 <span class="sc">==</span> <span class="st">&quot;HSOK&quot;</span>],</span>
<span id="cb31-30"><a href="#cb31-30" aria-hidden="true" tabindex="-1"></a>                     <span class="at">x =</span> df_kp<span class="sc">$</span>mid[df_kp<span class="sc">$</span>p2 <span class="sc">==</span> <span class="st">&quot;HSOK&quot;</span>],</span>
<span id="cb31-31"><a href="#cb31-31" aria-hidden="true" tabindex="-1"></a>                     <span class="at">y =</span> df_kp<span class="sc">$</span>mean_fd[df_kp<span class="sc">$</span>p2 <span class="sc">==</span> <span class="st">&quot;HSOK&quot;</span>],</span>
<span id="cb31-32"><a href="#cb31-32" aria-hidden="true" tabindex="-1"></a>                     <span class="at">col =</span> <span class="st">&quot;#631E68&quot;</span>,</span>
<span id="cb31-33"><a href="#cb31-33" aria-hidden="true" tabindex="-1"></a>                     <span class="at">r0=</span><span class="fl">0.3</span>, <span class="at">r1 =</span> <span class="dv">1</span>)</span>
<span id="cb31-34"><a href="#cb31-34" aria-hidden="true" tabindex="-1"></a><span class="co"># Add SNP density data</span></span>
<span id="cb31-35"><a href="#cb31-35" aria-hidden="true" tabindex="-1"></a><span class="fu">kpPlotDensity</span>(kp, <span class="at">data=</span>mikk_ranges, <span class="at">col =</span> <span class="st">&quot;#49A379&quot;</span>,</span>
<span id="cb31-36"><a href="#cb31-36" aria-hidden="true" tabindex="-1"></a>              <span class="at">r0=</span><span class="dv">0</span>, <span class="at">r1=</span><span class="fl">0.1</span>, </span>
<span id="cb31-37"><a href="#cb31-37" aria-hidden="true" tabindex="-1"></a>              <span class="at">window.size =</span> <span class="dv">25000</span>)</span>
<span id="cb31-38"><a href="#cb31-38" aria-hidden="true" tabindex="-1"></a><span class="fu">kpPlotDensity</span>(kp, <span class="at">data=</span>ol_ranges_list<span class="sc">$</span>hni, <span class="at">col =</span> <span class="st">&quot;#F6673A&quot;</span>,</span>
<span id="cb31-39"><a href="#cb31-39" aria-hidden="true" tabindex="-1"></a>              <span class="at">r0=</span><span class="fl">0.1</span>, <span class="at">r1=</span><span class="fl">0.2</span>, </span>
<span id="cb31-40"><a href="#cb31-40" aria-hidden="true" tabindex="-1"></a>              <span class="at">window.size =</span> <span class="dv">25000</span>)</span>
<span id="cb31-41"><a href="#cb31-41" aria-hidden="true" tabindex="-1"></a><span class="fu">kpPlotDensity</span>(kp, <span class="at">data=</span>ol_ranges_list<span class="sc">$</span>hsok, <span class="at">col =</span> <span class="st">&quot;#631E68&quot;</span>, </span>
<span id="cb31-42"><a href="#cb31-42" aria-hidden="true" tabindex="-1"></a>              <span class="at">r0=</span><span class="fl">0.2</span>, <span class="at">r1=</span><span class="fl">0.3</span>, </span>
<span id="cb31-43"><a href="#cb31-43" aria-hidden="true" tabindex="-1"></a>              <span class="at">window.size =</span> <span class="dv">25000</span>)</span>
<span id="cb31-44"><a href="#cb31-44" aria-hidden="true" tabindex="-1"></a><span class="co">#kpPlotDensity(kp, data=ol_ranges_list$hdrr, col = &quot;#F3B61F&quot;,</span></span>
<span id="cb31-45"><a href="#cb31-45" aria-hidden="true" tabindex="-1"></a><span class="co">#              r0=0.45, r1=0.6,</span></span>
<span id="cb31-46"><a href="#cb31-46" aria-hidden="true" tabindex="-1"></a><span class="co">#              window.size = 25000)</span></span>
<span id="cb31-47"><a href="#cb31-47" aria-hidden="true" tabindex="-1"></a><span class="co"># Add exon density to ideogram</span></span>
<span id="cb31-48"><a href="#cb31-48" aria-hidden="true" tabindex="-1"></a><span class="fu">kpPlotDensity</span>(kp, <span class="at">data=</span>ex_ranges, <span class="at">col =</span> <span class="st">&quot;#f77cb5&quot;</span>,</span>
<span id="cb31-49"><a href="#cb31-49" aria-hidden="true" tabindex="-1"></a>              <span class="at">data.panel =</span> <span class="st">&quot;ideogram&quot;</span>,</span>
<span id="cb31-50"><a href="#cb31-50" aria-hidden="true" tabindex="-1"></a>              <span class="at">window.size =</span> <span class="dv">25000</span>,</span>
<span id="cb31-51"><a href="#cb31-51" aria-hidden="true" tabindex="-1"></a>              <span class="at">r0 =</span> <span class="fl">0.5</span>, <span class="at">r1 =</span> <span class="dv">1</span>)</span>
<span id="cb31-52"><a href="#cb31-52" aria-hidden="true" tabindex="-1"></a><span class="fu">kpPlotDensity</span>(kp, <span class="at">data=</span>ex_ranges, <span class="at">col =</span> <span class="st">&quot;#f77cb5&quot;</span>,</span>
<span id="cb31-53"><a href="#cb31-53" aria-hidden="true" tabindex="-1"></a>              <span class="at">data.panel =</span> <span class="st">&quot;ideogram&quot;</span>,</span>
<span id="cb31-54"><a href="#cb31-54" aria-hidden="true" tabindex="-1"></a>              <span class="at">window.size =</span> <span class="dv">25000</span>,</span>
<span id="cb31-55"><a href="#cb31-55" aria-hidden="true" tabindex="-1"></a>              <span class="at">r0 =</span> <span class="fl">0.5</span>, <span class="at">r1 =</span> <span class="dv">0</span>)</span>
<span id="cb31-56"><a href="#cb31-56" aria-hidden="true" tabindex="-1"></a><span class="co"># Add labels</span></span>
<span id="cb31-57"><a href="#cb31-57" aria-hidden="true" tabindex="-1"></a><span class="fu">kpAddLabels</span>(kp, <span class="at">labels=</span><span class="st">&quot;MIKK&quot;</span>,</span>
<span id="cb31-58"><a href="#cb31-58" aria-hidden="true" tabindex="-1"></a>            <span class="at">r0=</span><span class="dv">0</span>, <span class="at">r1=</span><span class="fl">0.05</span>,</span>
<span id="cb31-59"><a href="#cb31-59" aria-hidden="true" tabindex="-1"></a>            <span class="at">label.margin =</span> <span class="fl">0.001</span>,</span>
<span id="cb31-60"><a href="#cb31-60" aria-hidden="true" tabindex="-1"></a>            <span class="at">cex =</span> <span class="fl">0.4</span>)</span>
<span id="cb31-61"><a href="#cb31-61" aria-hidden="true" tabindex="-1"></a><span class="fu">kpAddLabels</span>(kp, <span class="at">labels=</span><span class="st">&quot;HNI&quot;</span>,</span>
<span id="cb31-62"><a href="#cb31-62" aria-hidden="true" tabindex="-1"></a>            <span class="at">r0=</span><span class="fl">0.1</span>, <span class="at">r1=</span><span class="fl">0.15</span>, </span>
<span id="cb31-63"><a href="#cb31-63" aria-hidden="true" tabindex="-1"></a>            <span class="at">label.margin =</span> <span class="fl">0.001</span>,</span>
<span id="cb31-64"><a href="#cb31-64" aria-hidden="true" tabindex="-1"></a>            <span class="at">cex =</span> <span class="fl">0.4</span>)</span>
<span id="cb31-65"><a href="#cb31-65" aria-hidden="true" tabindex="-1"></a><span class="fu">kpAddLabels</span>(kp, <span class="at">labels=</span><span class="st">&quot;HSOK&quot;</span>,</span>
<span id="cb31-66"><a href="#cb31-66" aria-hidden="true" tabindex="-1"></a>            <span class="at">r0=</span><span class="fl">0.2</span>, <span class="at">r1=</span><span class="fl">0.25</span>,</span>
<span id="cb31-67"><a href="#cb31-67" aria-hidden="true" tabindex="-1"></a>            <span class="at">label.margin =</span> <span class="fl">0.001</span>,</span>
<span id="cb31-68"><a href="#cb31-68" aria-hidden="true" tabindex="-1"></a>            <span class="at">cex =</span> <span class="fl">0.4</span>)</span>
<span id="cb31-69"><a href="#cb31-69" aria-hidden="true" tabindex="-1"></a><span class="co">#kpAddLabels(kp, labels=&quot;HdrR&quot;,</span></span>
<span id="cb31-70"><a href="#cb31-70" aria-hidden="true" tabindex="-1"></a><span class="co">#            r0=0.45, r1=0.6, </span></span>
<span id="cb31-71"><a href="#cb31-71" aria-hidden="true" tabindex="-1"></a><span class="co">#            cex = 0.4)</span></span>
<span id="cb31-72"><a href="#cb31-72" aria-hidden="true" tabindex="-1"></a><span class="fu">kpAddLabels</span>(kp, <span class="at">labels=</span><span class="fu">bquote</span>(<span class="fu">italic</span>(f[d])),</span>
<span id="cb31-73"><a href="#cb31-73" aria-hidden="true" tabindex="-1"></a>            <span class="at">r0=</span><span class="fl">0.3</span>, <span class="at">r1=</span><span class="dv">1</span>, </span>
<span id="cb31-74"><a href="#cb31-74" aria-hidden="true" tabindex="-1"></a>            <span class="at">label.margin =</span> <span class="fl">0.035</span>,</span>
<span id="cb31-75"><a href="#cb31-75" aria-hidden="true" tabindex="-1"></a>            <span class="at">cex =</span> <span class="fl">0.6</span>)</span>
<span id="cb31-76"><a href="#cb31-76" aria-hidden="true" tabindex="-1"></a><span class="fu">dev.off</span>()</span></code></pre></div>
<div class="sourceCode" id="cb32"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">include_graphics</span>(<span class="fu">basename</span>(file_out))</span></code></pre></div>
<p><img src="20210318_fd_with_density_all.png" width="4250" /></p>
</div>
<div id="chromosome-4" class="section level4" number="4.2.2.6">
<h4 number="4.2.2.6"><span class="header-section-number">4.2.2.6</span> Chromosome 4</h4>
<div class="sourceCode" id="cb33"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a>out_file <span class="ot">=</span> <span class="fu">file.path</span>(plots_dir, <span class="st">&quot;20210318_fd_with_density_chr_4.png&quot;</span>)</span></code></pre></div>
<div class="sourceCode" id="cb34"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="fu">png</span>(<span class="at">file=</span>out_file,</span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">width=</span><span class="dv">5500</span>,</span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">height=</span><span class="dv">1186</span>,</span>
<span id="cb34-4"><a href="#cb34-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">units =</span> <span class="st">&quot;px&quot;</span>,</span>
<span id="cb34-5"><a href="#cb34-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">res =</span> <span class="dv">400</span>)</span>
<span id="cb34-6"><a href="#cb34-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-7"><a href="#cb34-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot </span></span>
<span id="cb34-8"><a href="#cb34-8" aria-hidden="true" tabindex="-1"></a>kp <span class="ot">=</span> <span class="fu">plotKaryotype</span>(med_genome, <span class="at">chromosomes =</span> <span class="st">&quot;4&quot;</span>, <span class="at">cex =</span> <span class="fl">1.5</span>)</span>
<span id="cb34-9"><a href="#cb34-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Add base numbers </span></span>
<span id="cb34-10"><a href="#cb34-10" aria-hidden="true" tabindex="-1"></a>karyoploteR<span class="sc">::</span><span class="fu">kpAddBaseNumbers</span>(kp, <span class="at">tick.dist =</span> <span class="dv">5000000</span>, <span class="at">cex =</span> <span class="fl">0.7</span>)</span>
<span id="cb34-11"><a href="#cb34-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Add data backgrounds</span></span>
<span id="cb34-12"><a href="#cb34-12" aria-hidden="true" tabindex="-1"></a>karyoploteR<span class="sc">::</span><span class="fu">kpDataBackground</span>(kp, <span class="at">r0=</span><span class="dv">0</span>, <span class="at">r1 =</span> <span class="dv">1</span>, <span class="at">color =</span> <span class="st">&quot;white&quot;</span>)</span>
<span id="cb34-13"><a href="#cb34-13" aria-hidden="true" tabindex="-1"></a><span class="co"># Add axis label</span></span>
<span id="cb34-14"><a href="#cb34-14" aria-hidden="true" tabindex="-1"></a><span class="fu">kpAxis</span>(kp, <span class="at">r0=</span><span class="fl">0.3</span>, <span class="at">r1 =</span> <span class="dv">1</span>, <span class="at">cex =</span> <span class="fl">0.8</span>)</span>
<span id="cb34-15"><a href="#cb34-15" aria-hidden="true" tabindex="-1"></a><span class="co"># Add fd data</span></span>
<span id="cb34-16"><a href="#cb34-16" aria-hidden="true" tabindex="-1"></a>lwd <span class="ot">=</span> <span class="dv">2</span></span>
<span id="cb34-17"><a href="#cb34-17" aria-hidden="true" tabindex="-1"></a>karyoploteR<span class="sc">::</span><span class="fu">kpLines</span>(kp,</span>
<span id="cb34-18"><a href="#cb34-18" aria-hidden="true" tabindex="-1"></a>                     <span class="at">chr =</span> df_kp<span class="sc">$</span>scaffold[df_kp<span class="sc">$</span>p2 <span class="sc">==</span> <span class="st">&quot;HNI&quot;</span>],</span>
<span id="cb34-19"><a href="#cb34-19" aria-hidden="true" tabindex="-1"></a>                     <span class="at">x =</span> df_kp<span class="sc">$</span>mid[df_kp<span class="sc">$</span>p2 <span class="sc">==</span> <span class="st">&quot;HNI&quot;</span>],</span>
<span id="cb34-20"><a href="#cb34-20" aria-hidden="true" tabindex="-1"></a>                     <span class="at">y =</span> df_kp<span class="sc">$</span>mean_fd[df_kp<span class="sc">$</span>p2 <span class="sc">==</span> <span class="st">&quot;HNI&quot;</span>],</span>
<span id="cb34-21"><a href="#cb34-21" aria-hidden="true" tabindex="-1"></a>                     <span class="at">col =</span> <span class="st">&quot;#F6673A&quot;</span>,</span>
<span id="cb34-22"><a href="#cb34-22" aria-hidden="true" tabindex="-1"></a>                     <span class="at">r0=</span><span class="fl">0.3</span>, <span class="at">r1 =</span> <span class="dv">1</span>,</span>
<span id="cb34-23"><a href="#cb34-23" aria-hidden="true" tabindex="-1"></a>                     <span class="at">lwd =</span> lwd)</span>
<span id="cb34-24"><a href="#cb34-24" aria-hidden="true" tabindex="-1"></a>karyoploteR<span class="sc">::</span><span class="fu">kpLines</span>(kp,</span>
<span id="cb34-25"><a href="#cb34-25" aria-hidden="true" tabindex="-1"></a>                     <span class="at">chr =</span> df_kp<span class="sc">$</span>scaffold[df_kp<span class="sc">$</span>p2 <span class="sc">==</span> <span class="st">&quot;HdrR&quot;</span>],</span>
<span id="cb34-26"><a href="#cb34-26" aria-hidden="true" tabindex="-1"></a>                     <span class="at">x =</span> df_kp<span class="sc">$</span>mid[df_kp<span class="sc">$</span>p2 <span class="sc">==</span> <span class="st">&quot;HdrR&quot;</span>],</span>
<span id="cb34-27"><a href="#cb34-27" aria-hidden="true" tabindex="-1"></a>                     <span class="at">y =</span> df_kp<span class="sc">$</span>mean_fd[df_kp<span class="sc">$</span>p2 <span class="sc">==</span> <span class="st">&quot;HdrR&quot;</span>],</span>
<span id="cb34-28"><a href="#cb34-28" aria-hidden="true" tabindex="-1"></a>                     <span class="at">col =</span> <span class="st">&quot;#F3B61F&quot;</span>,</span>
<span id="cb34-29"><a href="#cb34-29" aria-hidden="true" tabindex="-1"></a>                     <span class="at">r0=</span><span class="fl">0.3</span>, <span class="at">r1 =</span> <span class="dv">1</span>,</span>
<span id="cb34-30"><a href="#cb34-30" aria-hidden="true" tabindex="-1"></a>                     <span class="at">lwd =</span> lwd)</span>
<span id="cb34-31"><a href="#cb34-31" aria-hidden="true" tabindex="-1"></a>karyoploteR<span class="sc">::</span><span class="fu">kpLines</span>(kp,</span>
<span id="cb34-32"><a href="#cb34-32" aria-hidden="true" tabindex="-1"></a>                     <span class="at">chr =</span> df_kp<span class="sc">$</span>scaffold[df_kp<span class="sc">$</span>p2 <span class="sc">==</span> <span class="st">&quot;HSOK&quot;</span>],</span>
<span id="cb34-33"><a href="#cb34-33" aria-hidden="true" tabindex="-1"></a>                     <span class="at">x =</span> df_kp<span class="sc">$</span>mid[df_kp<span class="sc">$</span>p2 <span class="sc">==</span> <span class="st">&quot;HSOK&quot;</span>],</span>
<span id="cb34-34"><a href="#cb34-34" aria-hidden="true" tabindex="-1"></a>                     <span class="at">y =</span> df_kp<span class="sc">$</span>mean_fd[df_kp<span class="sc">$</span>p2 <span class="sc">==</span> <span class="st">&quot;HSOK&quot;</span>],</span>
<span id="cb34-35"><a href="#cb34-35" aria-hidden="true" tabindex="-1"></a>                     <span class="at">col =</span> <span class="st">&quot;#631E68&quot;</span>,</span>
<span id="cb34-36"><a href="#cb34-36" aria-hidden="true" tabindex="-1"></a>                     <span class="at">r0=</span><span class="fl">0.3</span>, <span class="at">r1 =</span> <span class="dv">1</span>,</span>
<span id="cb34-37"><a href="#cb34-37" aria-hidden="true" tabindex="-1"></a>                     <span class="at">lwd =</span> lwd)</span>
<span id="cb34-38"><a href="#cb34-38" aria-hidden="true" tabindex="-1"></a><span class="co"># Add SNP density data</span></span>
<span id="cb34-39"><a href="#cb34-39" aria-hidden="true" tabindex="-1"></a><span class="fu">kpPlotDensity</span>(kp, <span class="at">data=</span>mikk_ranges, <span class="at">col =</span> <span class="st">&quot;#49A379&quot;</span>,</span>
<span id="cb34-40"><a href="#cb34-40" aria-hidden="true" tabindex="-1"></a>              <span class="at">r0=</span><span class="dv">0</span>, <span class="at">r1=</span><span class="fl">0.1</span>, </span>
<span id="cb34-41"><a href="#cb34-41" aria-hidden="true" tabindex="-1"></a>              <span class="at">window.size =</span> <span class="dv">25000</span>)</span>
<span id="cb34-42"><a href="#cb34-42" aria-hidden="true" tabindex="-1"></a><span class="fu">kpPlotDensity</span>(kp, <span class="at">data=</span>ol_ranges_list<span class="sc">$</span>hni, <span class="at">col =</span> <span class="st">&quot;#F6673A&quot;</span>,</span>
<span id="cb34-43"><a href="#cb34-43" aria-hidden="true" tabindex="-1"></a>              <span class="at">r0=</span><span class="fl">0.1</span>, <span class="at">r1=</span><span class="fl">0.2</span>, </span>
<span id="cb34-44"><a href="#cb34-44" aria-hidden="true" tabindex="-1"></a>              <span class="at">window.size =</span> <span class="dv">25000</span>)</span>
<span id="cb34-45"><a href="#cb34-45" aria-hidden="true" tabindex="-1"></a><span class="fu">kpPlotDensity</span>(kp, <span class="at">data=</span>ol_ranges_list<span class="sc">$</span>hsok, <span class="at">col =</span> <span class="st">&quot;#631E68&quot;</span>, </span>
<span id="cb34-46"><a href="#cb34-46" aria-hidden="true" tabindex="-1"></a>              <span class="at">r0=</span><span class="fl">0.2</span>, <span class="at">r1=</span><span class="fl">0.3</span>, </span>
<span id="cb34-47"><a href="#cb34-47" aria-hidden="true" tabindex="-1"></a>              <span class="at">window.size =</span> <span class="dv">25000</span>)</span>
<span id="cb34-48"><a href="#cb34-48" aria-hidden="true" tabindex="-1"></a><span class="co">#kpPlotDensity(kp, data=ol_ranges_list$hdrr, col = &quot;#F3B61F&quot;,</span></span>
<span id="cb34-49"><a href="#cb34-49" aria-hidden="true" tabindex="-1"></a><span class="co">#              r0=0.45, r1=0.6,</span></span>
<span id="cb34-50"><a href="#cb34-50" aria-hidden="true" tabindex="-1"></a><span class="co">#              window.size = 25000)</span></span>
<span id="cb34-51"><a href="#cb34-51" aria-hidden="true" tabindex="-1"></a><span class="co"># Add exon density to ideogram</span></span>
<span id="cb34-52"><a href="#cb34-52" aria-hidden="true" tabindex="-1"></a><span class="fu">kpPlotDensity</span>(kp, <span class="at">data=</span>ex_ranges, <span class="at">col =</span> <span class="st">&quot;#f77cb5&quot;</span>,</span>
<span id="cb34-53"><a href="#cb34-53" aria-hidden="true" tabindex="-1"></a>              <span class="at">data.panel =</span> <span class="st">&quot;ideogram&quot;</span>,</span>
<span id="cb34-54"><a href="#cb34-54" aria-hidden="true" tabindex="-1"></a>              <span class="at">window.size =</span> <span class="dv">25000</span>,</span>
<span id="cb34-55"><a href="#cb34-55" aria-hidden="true" tabindex="-1"></a>              <span class="at">r0 =</span> <span class="fl">0.5</span>, <span class="at">r1 =</span> <span class="dv">1</span>)</span>
<span id="cb34-56"><a href="#cb34-56" aria-hidden="true" tabindex="-1"></a><span class="fu">kpPlotDensity</span>(kp, <span class="at">data=</span>ex_ranges, <span class="at">col =</span> <span class="st">&quot;#f77cb5&quot;</span>,</span>
<span id="cb34-57"><a href="#cb34-57" aria-hidden="true" tabindex="-1"></a>              <span class="at">data.panel =</span> <span class="st">&quot;ideogram&quot;</span>,</span>
<span id="cb34-58"><a href="#cb34-58" aria-hidden="true" tabindex="-1"></a>              <span class="at">window.size =</span> <span class="dv">25000</span>,</span>
<span id="cb34-59"><a href="#cb34-59" aria-hidden="true" tabindex="-1"></a>              <span class="at">r0 =</span> <span class="fl">0.5</span>, <span class="at">r1 =</span> <span class="dv">0</span>)</span>
<span id="cb34-60"><a href="#cb34-60" aria-hidden="true" tabindex="-1"></a><span class="co"># Add labels</span></span>
<span id="cb34-61"><a href="#cb34-61" aria-hidden="true" tabindex="-1"></a><span class="fu">kpAddLabels</span>(kp, <span class="at">labels=</span><span class="st">&quot;MIKK&quot;</span>,</span>
<span id="cb34-62"><a href="#cb34-62" aria-hidden="true" tabindex="-1"></a>            <span class="at">r0=</span><span class="dv">0</span>, <span class="at">r1=</span><span class="fl">0.05</span>,</span>
<span id="cb34-63"><a href="#cb34-63" aria-hidden="true" tabindex="-1"></a>            <span class="at">label.margin =</span> <span class="fl">0.001</span>,</span>
<span id="cb34-64"><a href="#cb34-64" aria-hidden="true" tabindex="-1"></a>            <span class="at">cex =</span> <span class="fl">0.5</span>)</span>
<span id="cb34-65"><a href="#cb34-65" aria-hidden="true" tabindex="-1"></a><span class="fu">kpAddLabels</span>(kp, <span class="at">labels=</span><span class="st">&quot;HNI&quot;</span>,</span>
<span id="cb34-66"><a href="#cb34-66" aria-hidden="true" tabindex="-1"></a>            <span class="at">r0=</span><span class="fl">0.1</span>, <span class="at">r1=</span><span class="fl">0.15</span>, </span>
<span id="cb34-67"><a href="#cb34-67" aria-hidden="true" tabindex="-1"></a>            <span class="at">label.margin =</span> <span class="fl">0.001</span>,</span>
<span id="cb34-68"><a href="#cb34-68" aria-hidden="true" tabindex="-1"></a>            <span class="at">cex =</span> <span class="fl">0.5</span>)</span>
<span id="cb34-69"><a href="#cb34-69" aria-hidden="true" tabindex="-1"></a><span class="fu">kpAddLabels</span>(kp, <span class="at">labels=</span><span class="st">&quot;HSOK&quot;</span>,</span>
<span id="cb34-70"><a href="#cb34-70" aria-hidden="true" tabindex="-1"></a>            <span class="at">r0=</span><span class="fl">0.2</span>, <span class="at">r1=</span><span class="fl">0.25</span>,</span>
<span id="cb34-71"><a href="#cb34-71" aria-hidden="true" tabindex="-1"></a>            <span class="at">label.margin =</span> <span class="fl">0.001</span>,</span>
<span id="cb34-72"><a href="#cb34-72" aria-hidden="true" tabindex="-1"></a>            <span class="at">cex =</span> <span class="fl">0.5</span>)</span>
<span id="cb34-73"><a href="#cb34-73" aria-hidden="true" tabindex="-1"></a><span class="co">#kpAddLabels(kp, labels=&quot;HdrR&quot;,</span></span>
<span id="cb34-74"><a href="#cb34-74" aria-hidden="true" tabindex="-1"></a><span class="co">#            r0=0.45, r1=0.6, </span></span>
<span id="cb34-75"><a href="#cb34-75" aria-hidden="true" tabindex="-1"></a><span class="co">#            cex = 0.4)</span></span>
<span id="cb34-76"><a href="#cb34-76" aria-hidden="true" tabindex="-1"></a><span class="fu">kpAddLabels</span>(kp, <span class="at">labels=</span><span class="fu">bquote</span>(<span class="fu">italic</span>(f[d])),</span>
<span id="cb34-77"><a href="#cb34-77" aria-hidden="true" tabindex="-1"></a>            <span class="at">r0=</span><span class="fl">0.3</span>, <span class="at">r1=</span><span class="dv">1</span>, </span>
<span id="cb34-78"><a href="#cb34-78" aria-hidden="true" tabindex="-1"></a>            <span class="at">label.margin =</span> <span class="fl">0.035</span>,</span>
<span id="cb34-79"><a href="#cb34-79" aria-hidden="true" tabindex="-1"></a>            <span class="at">cex =</span> <span class="dv">1</span>)</span>
<span id="cb34-80"><a href="#cb34-80" aria-hidden="true" tabindex="-1"></a><span class="fu">dev.off</span>()</span></code></pre></div>
<div class="sourceCode" id="cb35"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">include_graphics</span>(<span class="fu">basename</span>(out_file))</span></code></pre></div>
<p><img src="20210318_fd_with_density_chr_4.png" width="2750" /></p>
</div>
</div>
</div>
</div>
<div id="final-figure" class="section level1" number="5">
<h1 number="5"><span class="header-section-number">5</span> Final figure</h1>
<div id="abba-baba-diagram" class="section level2" number="5.1">
<h2 number="5.1"><span class="header-section-number">5.1</span> ABBA BABA diagram</h2>
<p>Created with <a href="https://vectr.com/">Vectr</a> and saved here: <code>plots/introgression/20210318_abba_diagram.svg</code></p>
</div>
<div id="compile-all" class="section level2" number="5.2">
<h2 number="5.2"><span class="header-section-number">5.2</span> Compile all</h2>
<div class="sourceCode" id="cb36"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a>abba_diagram <span class="ot">=</span> here<span class="sc">::</span><span class="fu">here</span>(plots_dir, <span class="st">&quot;abba_diagram.png&quot;</span>)</span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a>tree <span class="ot">=</span> here<span class="sc">::</span><span class="fu">here</span>(plots_dir, <span class="st">&quot;tree_oryzias_with_ancestor.png&quot;</span>)</span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a>karyo_chr4 <span class="ot">=</span> here<span class="sc">::</span><span class="fu">here</span>(plots_dir, <span class="st">&quot;20210318_fd_with_density_chr_4.png&quot;</span>)</span>
<span id="cb36-4"><a href="#cb36-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-5"><a href="#cb36-5" aria-hidden="true" tabindex="-1"></a>final_abba <span class="ot">=</span> <span class="fu">ggdraw</span>() <span class="sc">+</span></span>
<span id="cb36-6"><a href="#cb36-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">draw_image</span>(tree,</span>
<span id="cb36-7"><a href="#cb36-7" aria-hidden="true" tabindex="-1"></a>            <span class="at">x =</span> <span class="dv">0</span>, <span class="at">y =</span> .<span class="dv">7</span>, <span class="at">width =</span> .<span class="dv">4</span>, <span class="at">height =</span> .<span class="dv">35</span>, <span class="at">vjust =</span> .<span class="dv">1</span>, <span class="at">hjust =</span> <span class="sc">-</span>.<span class="dv">1</span>, <span class="at">scale =</span> <span class="fl">1.2</span>) <span class="sc">+</span></span>
<span id="cb36-8"><a href="#cb36-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">draw_image</span>(karyo_chr4,</span>
<span id="cb36-9"><a href="#cb36-9" aria-hidden="true" tabindex="-1"></a>             <span class="at">x =</span> <span class="dv">0</span>, <span class="at">y =</span> <span class="dv">0</span>, <span class="at">width =</span> <span class="dv">1</span>, <span class="at">height =</span> <span class="fl">0.3</span>, <span class="at">scale =</span> <span class="fl">1.2</span>) <span class="sc">+</span>   </span>
<span id="cb36-10"><a href="#cb36-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">draw_plot</span>(fstat_plot,</span>
<span id="cb36-11"><a href="#cb36-11" aria-hidden="true" tabindex="-1"></a>             <span class="at">x =</span> .<span class="dv">4</span>, <span class="at">y =</span> .<span class="dv">3</span>, <span class="at">width =</span> .<span class="dv">6</span>, <span class="at">height =</span> .<span class="dv">7</span>) <span class="sc">+</span></span>
<span id="cb36-12"><a href="#cb36-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">draw_image</span>(abba_diagram,</span>
<span id="cb36-13"><a href="#cb36-13" aria-hidden="true" tabindex="-1"></a>          <span class="at">x =</span> <span class="dv">0</span>, <span class="at">y =</span> .<span class="dv">3</span>, <span class="at">width =</span> .<span class="dv">4</span>, <span class="at">height =</span> .<span class="dv">35</span>, <span class="at">vjust =</span> <span class="sc">-</span>.<span class="dv">05</span>, <span class="at">scale =</span> <span class="fl">1.1</span>) <span class="sc">+</span></span>
<span id="cb36-14"><a href="#cb36-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">draw_plot_label</span>(<span class="at">label =</span> <span class="fu">c</span>(<span class="st">&quot;A&quot;</span>, <span class="st">&quot;B&quot;</span>, <span class="st">&quot;C&quot;</span>, <span class="st">&quot;D&quot;</span>), <span class="at">size =</span> <span class="dv">16</span>,</span>
<span id="cb36-15"><a href="#cb36-15" aria-hidden="true" tabindex="-1"></a>                  <span class="at">x =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>, .<span class="dv">38</span>, <span class="dv">0</span>), <span class="at">y =</span> <span class="fu">c</span>(<span class="dv">1</span>, .<span class="dv">7</span>, <span class="dv">1</span>, .<span class="dv">3</span>))  </span>
<span id="cb36-16"><a href="#cb36-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-17"><a href="#cb36-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-18"><a href="#cb36-18" aria-hidden="true" tabindex="-1"></a>final_abba</span></code></pre></div>
<p><img src="20210318_introgression_files/figure-html/unnamed-chunk-45-1.svg" width="1344" /></p>
<div class="sourceCode" id="cb37"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(here<span class="sc">::</span><span class="fu">here</span>(plots_dir, <span class="st">&quot;20210318_final_figure.png&quot;</span>),</span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a>       <span class="at">width =</span> <span class="dv">35</span>,</span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a>       <span class="at">height =</span> <span class="fl">21.875</span>,</span>
<span id="cb37-4"><a href="#cb37-4" aria-hidden="true" tabindex="-1"></a>       <span class="at">units =</span> <span class="st">&quot;cm&quot;</span>,</span>
<span id="cb37-5"><a href="#cb37-5" aria-hidden="true" tabindex="-1"></a>       <span class="at">dpi =</span> <span class="dv">500</span>)</span></code></pre></div>
</div>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
